---
title: "DEseq"
author: "Kate Petrenko"
date: "2024-07-16"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = FALSE,
	message = FALSE,
	warning = FALSE
)
library(data.table)
library(Seurat)
library(scDblFinder)
library(DoubletFinder)
library(dplyr)
library(remotes)
# remotes::install_github("carmonalab/STACAS")
library(STACAS)
library(SingleR)
# install.packages("scGate")
library(scGate)
library(ggplot2)
library(BiocParallel)
library(harmony)
library(RColorBrewer)
library(stringr)
library(cowplot)
library(scuttle)
# install.packages("devtools")
# BiocManager::install("Homo.sapiens")
# BiocManager::install("jpmam1/scalop") #forked copy of a really good guy 
# devtools::install_github("jlaffy/infercna")
library(infercna)
library(infercnv)
library(parallel)
library(future)
library(clustree)
library(dittoSeq)
# install.packages("ggh4x")
library(ggh4x)
#BiocManager::install('EnhancedVolcano')
library(EnhancedVolcano)
library(clusterProfiler)
library("org.Hs.eg.db")
library(ReactomePA)
library(msigdbr)
# install.packages('qs')
library(qs)
library(DESeq2)
library(magrittr)
library(tibble)
library(tidyr)
library(pheatmap)
library(edgeR)
library(limma)
# BiocManager::install('apeglm')
# install.packages('ashr')
# install.packages('VennDiagram')
library(VennDiagram)
# BiocManager::install("glmGamPoi")
library(glmGamPoi)
# BiocManager::install("metapod")
library(metapod)
# install.packages('forestplot')
library(forestplot)
```

```{r global variables, include=FALSE}
set.seed(100)
DATADIR = '~/neuroblastoma/data/'
WORKDIR = '~/neuroblastoma/'
PREPRDATADIR = '~/neuroblastoma/preprocessing_data/'

my_colors <- c("#9ACD32FF", "#4682B4FF",  "#DDA0DDFF", "#FFA07AFF", "#8FBC8BFF",
               "#40E0D0FF",  "#F0E68CFF", "#5F9EA0FF", "#D2B48CFF",  
               "#6495EDFF", "#FF69B4FF", "#BA55D3FF",  "#F08080FF", "#32CD32FF",  
               "#FFDAB9FF",  "#87CEEBFF")
```

```{r read data}
seu <- qread( paste0(PREPRDATADIR, 'seu_integrated_mypipelineannot_v2.qs'))
bulk <- qread(paste0(PREPRDATADIR, 'bulk_replicates_v1.qs'))
t_cells <- subset(seu, ANNOTATION_FINAL == 'T_cells')
b_cells <- subset(seu, ANNOTATION_FINAL == 'B_cells')
myeloid_cells <- subset(seu, ANNOTATION_FINAL == 'myeloid_cells')
```

```{r eval=FALSE, include=FALSE}
seu <- qread( paste0(PREPRDATADIR, 'seu_integrated_mypipelineannot_v1.qs'))

seu <- seu %>% 
  RunHarmony(., 'SampleID',
             lambda = 1, verbose = FALSE) %>% 
  RunUMAP(reduction = "harmony", dims = 1:20, verbose=F) 

seu$ANNOTATION_FINAL <- ifelse(seu$ANNOTATION_FINAL == "endothelial_cells", "stromal_cells", seu$ANNOTATION_FINAL)
seu <- subset(seu, ANNOTATION_FINAL %in% c("myeloid_cells", "B_cells", "T_cells",
                                           "malignant_cells","stromal_cells"))

qsave(seu, paste0(PREPRDATADIR, 'seu_integrated_mypipelineannot_v2.qs'))
```

```{r create pseudoreplicates, eval=FALSE, include=FALSE}
seu$less18M <- as.character(seu$less18M)
seu$rep <- NA
create_pseudorep <- function(seu) {
  
  rep1 <- c()
  rep2 <- c()
  rep3 <- c()
  reps <- list(rep1, rep2, rep3)
  
  samples <- seu$SampleID %>% unique
  cell_types <- seu$ANNOTATION_FINAL %>% unique
  sample_annot_table <- table(seu$ANNOTATION_FINAL, seu$SampleID)
  for (sample in samples) {
    for (celltype in cell_types) {
      
      if (sample_annot_table[celltype, sample] >= 150) {
        
        cells <- WhichCells(seu, expression = ANNOTATION_FINAL ==  celltype & SampleID == sample)
        cells_split <- split(cells,  cut(seq_along(cells), 3, labels = FALSE))
        reps[[1]] <- c(reps[[1]], cells_split[1])
        reps[[2]] <- c(reps[[2]], cells_split[2])
        reps[[3]] <- c(reps[[3]], cells_split[3])
      }
      else if (sample_annot_table[celltype, sample] >= 90) {
        cells <- WhichCells(seu, expression = ANNOTATION_FINAL ==  celltype & SampleID == sample)
        cells_split <- split(cells,  cut(seq_along(cells), 2, labels = FALSE))
        reps[[1]] <- c(reps[[1]], cells_split[1])
        reps[[2]] <- c(reps[[2]], cells_split[2])
      }
    }
  }
  return(reps)
}

replicates_list <- create_pseudorep(seu)

for (i in seq_along(replicates_list)) {
  replicate_name <- paste0("rep", i)
  cells <- unlist(replicates_list[[i]])
  seu$rep[cells] <- replicate_name
}
```

# 1 variant to create pseudobulk

```{r eval=FALSE, include=FALSE}
seu_bulk <- subset(seu, rep %in% c('rep1', 'rep2', 'rep3'))
bulk <- AggregateExpression(seu_bulk, return.seurat = T, slot = "counts", assays = "RNA", 
                            group.by = c("ANNOTATION_FINAL", "SampleID", "less18M", 'rep'))
Idents(bulk) <- "less18M"

qsave(bulk, paste0(PREPRDATADIR, 'bulk_replicates_v1.qs'))
```

# 2 variant to create pseudobulk
```{r eval=FALSE, include=FALSE}
seu_bulk_v2 <- as.SingleCellExperiment(subset(seu, rep %in% c('rep1', 'rep2', 'rep3')))
reduced_sce_b <- pseudobulk(subset(seu_bulk_v2, ANNOTATION_FINAL == "B_cells"), 
                            group_by = vars(condition = less18M, SampleID, rep), 
                          n_cells = n())
reduced_sce_t <- pseudobulk(subset(seu_bulk_v2, ANNOTATION_FINAL == "T_cells"), 
                            group_by = vars(condition = less18M, SampleID, rep), 
                          n_cells = n())
reduced_sce_m <- pseudobulk(subset(seu_bulk_v2, ANNOTATION_FINAL == "myeloid_cells"), 
                            group_by = vars(condition = less18M, SampleID, rep), 
                          n_cells = n())
reduced_sce_tumor <- pseudobulk(subset(seu_bulk_v2, ANNOTATION_FINAL == "malignant_cells"), 
                            group_by = vars(condition = less18M, SampleID, rep), 
                          n_cells = n())
reduced_sce_s <- pseudobulk(subset(seu_bulk_v2, ANNOTATION_FINAL == "stromal_cells"), 
                            group_by = vars(condition = less18M, SampleID, rep), 
                          n_cells = n())

fit <- glm_gp(reduced_sce_b, design = ~ condition + SampleID, size_factor = "ratio", verbose = TRUE)
```

# Create matrices for DEseq

```{r create matrices for deseq}
# bulk <- subset(bulk, Study != 'jansky')
create_count_matrix_from_bulk <- function(bulk) {
  cts <- bulk[["RNA"]]$counts %>% as.matrix()

  coldata <- bulk@meta.data 
  
  coldata$Study <- sapply(str_split(coldata$SampleID, "-"), function(x) {
    if (length(x) == 2) {
      return(x[2])
    } else if (length(x) == 3) {
      return(x[3])
    } else {
      return(NA)
    }
  })
  
  coldata$Technology <- '10x'
  coldata$Technology[coldata$Study == 'yuan'] <- 'Smart-seq2'
  
  ## SPLIT MATRIX BY CELL TYPE
  # transpose
  cts.t <- t(cts)
  # convert to data.frame
  cts.t <- as.data.frame(cts.t)
  # get values where to split
  splitRows <- gsub('_.*', '', rownames(cts.t))
  # split data.frame
  cts.split <- split.data.frame(cts.t,
                   f = factor(splitRows))
  
  # fix colnames and transpose
  cts.split.modified <- lapply(cts.split, function(x){
    rownames(x) <- rownames(x) %>% str_extract("[A-Za-z0-9-]+_(.*)", group = 1)
    t(x)
  
  })
  return(list(count_matrix = cts.split.modified, metadata = coldata))
}

cts.split.modified <- create_count_matrix_from_bulk(bulk)
coldata <- cts.split.modified$metadata
cts.split.modified <- cts.split.modified$count_matrix
```

# DEseq 

```{r function for deseq}
deseq_batch_correct <- function(list_matrix, coldata, cell_type, design) {
  
  count_matrix <- as.matrix(list_matrix[cell_type][[1]])
  coldata_cell <- coldata %>% dplyr::filter(ANNOTATION_FINAL == cell_type) %>%
                                set_rownames(rownames(.) %>% str_extract("[A-Za-z0-9-]+_(.*)",
                                                                         group = 1))
  dds <- DESeqDataSetFromMatrix(count_matrix, 
                              colData = coldata_cell, 
                              design = design)
  
  dds <- DESeq(dds)
  
  res <- results(dds,
                contrast= list('less18M_TRUE_vs_FALSE'))
  res <- lfcShrink(dds,
                   contrast= list('less18M_TRUE_vs_FALSE'),
                   type='ashr',
                   res=res)

  res_tbl <- res %>%
        data.frame() %>%
        rownames_to_column(var="gene") %>%
        as_tibble()
  
  normalized_counts <- counts(dds, 
                            normalized = TRUE)
  
  sig_res <- dplyr::filter(res_tbl, padj < 0.05) %>%
        dplyr::arrange(padj)
  
  sig_norm <- data.frame(normalized_counts) %>%
        rownames_to_column(var = "gene") %>%
        dplyr::filter(gene %in% sig_res$gene)

  # Set a color palette
  heat_colors <- brewer.pal(6, "RdBu")
  
  # Run pheatmap using the metadata data frame for the annotation
  p <- pheatmap(sig_norm[ , 2:length(colnames(sig_norm))], 
      color = heat_colors, 
      breaks = seq(-2, 2, length.out = 6),
      cluster_rows = T, 
      show_rownames = F,
      annotation = coldata_cell %>% 
        set_rownames(gsub('-', '.', rownames(.))) %>% dplyr::select(less18M), 
      border_color = NA, 
      fontsize = 10, 
      scale = "row", 
      silent=T,
      fontsize_row = 10, 
      height = 20)

  return(list(res = res_tbl, plot = p))
}
```

```{r run deseq}
b_cell_deseq_res <- deseq_batch_correct(cts.split.modified, coldata, 'B-cells', as.formula(' ~ Study + less18M'))
b_cells_plot <- b_cell_deseq_res$plot
b_cell_deseq_res <- b_cell_deseq_res$res

t_cell_deseq_res <- deseq_batch_correct(cts.split.modified, coldata, 'T-cells', as.formula(' ~ Study + less18M'))
t_cells_plot <- t_cell_deseq_res$plot
t_cell_deseq_res <- t_cell_deseq_res$res

myel_cell_deseq_res <- deseq_batch_correct(cts.split.modified, coldata, 'myeloid-cells', as.formula(' ~ Study + less18M'))
myel_cells_plot <- myel_cell_deseq_res$plot
myel_cell_deseq_res <- myel_cell_deseq_res$res
```

## Pseudobulk results

```{r fig.height=25, fig.width=18}
plot_grid(b_cells_plot[[4]], t_cells_plot[[4]], myel_cells_plot[[4]], nrow=3, 
          labels = c('A - B-cells', 'B - T-cells', 'C - Myeloid cells'))
```

# DEseq by study

```{r}
bulk$Study <- sapply(str_split(bulk$SampleID, "-"), function(x) {
  if (length(x) == 2) {
    return(x[2])
  } else if (length(x) == 3) {
    return(x[3])
  } else {
    return(NA)
  }
})

de_study <- function(study) {
  
  verh <- create_count_matrix_from_bulk(subset(bulk, Study == study))
  coldata2 <- verh$metadata
  verh <- verh$count_matrix
  
  if (study %in% c('fetahu', 'verhoeven')) {
    b_cell_deseq_verh <- deseq_batch_correct(verh, coldata2, 'B-cells', as.formula('~less18M'))
    b_cells_plot_verh <- b_cell_deseq_verh$plot
    b_cell_deseq_verh <- b_cell_deseq_verh$res
  }
  else {
    b_cells_plot_verh = ''
    b_cell_deseq_verh = ''
  }
  if (study %in% c('fetahu', 'verhoeven', 'dong')) {
    t_cell_deseq_verh <- deseq_batch_correct(verh, coldata2, 'T-cells', as.formula('~less18M'))
    t_cells_plot_verh <- t_cell_deseq_verh$plot
    t_cell_deseq_verh <- t_cell_deseq_verh$res
  }
  else {
    t_cells_plot_verh = ''
    t_cell_deseq_verh = ''
  }
  myel_cell_deseq_verh <- deseq_batch_correct(verh, coldata2, 'myeloid-cells',  as.formula('~less18M'))
  myel_cells_plot_verh <- myel_cell_deseq_verh$plot
  myel_cell_deseq_verh <- myel_cell_deseq_verh$res
  
  return(list(b_cell = list(plot = b_cells_plot_verh, res = b_cell_deseq_verh),
         t_cell = list(plot = t_cells_plot_verh, res = t_cell_deseq_verh),
         myeloid_cell = list(plot = myel_cells_plot_verh, res = myel_cell_deseq_verh)))
}
```


```{r run deseq by studies}
deseq_res_study <- mclapply(bulk$Study %>% unique, de_study, mc.cores = 5)
names(deseq_res_study) <- bulk$Study %>% unique
```

## Fetahu

```{r fig.height=25, fig.width=18}
plot_grid(deseq_res_study$fetahu$b_cell$plot[[4]], 
          deseq_res_study$fetahu$t_cell$plot[[4]], 
          deseq_res_study$fetahu$myeloid_cell$plot[[4]], 
          nrow=3, 
          labels = c('A - B-cells', 'B - T-cells', 'C - Myeloid cells'))
```

## Jansky

```{r fig.height=9, fig.width=18}
plot_grid(deseq_res_study$jansky$myeloid_cell$plot[[4]], 
          nrow=1, 
          labels = c('Myeloid cells'))
```

## Verhoeven

```{r fig.height=25, fig.width=18}
plot_grid(deseq_res_study$verhoeven$b_cell$plot[[4]], 
          deseq_res_study$verhoeven$t_cell$plot[[4]], 
          deseq_res_study$verhoeven$myeloid_cell$plot[[4]], 
          nrow=3, 
          labels = c('A - B-cells', 'B - T-cells', 'C - Myeloid cells'))
```

## Yuan

```{r fig.height=9, fig.width=18}
plot_grid(
          deseq_res_study$yuan$myeloid_cell$plot[[4]], 
          nrow=1, 
          labels = c('Myeloid cells'))
```

## Dong

```{r fig.height=14, fig.width=18}
plot_grid(
          deseq_res_study$dong$t_cell$plot[[4]], 
          deseq_res_study$dong$myeloid_cell$plot[[4]], 
          nrow=2, 
          labels = c('B - T-cells', 'C - Myeloid cells'))
```

## Intersections

```{r fig.height=5, fig.width=14}
myCol <- brewer.pal(5, "Pastel2")
b_cells_vienn <- venn.diagram(
        x = list(deseq_res_study$fetahu$b_cell$res %>% 
                   dplyr::filter(padj < 0.05) %>% .$gene, 
                 deseq_res_study$verhoeven$b_cell$res %>% 
                   dplyr::filter(padj < 0.05) %>% .$gene),
        category.names = c("Fetahu" , "Verhoeven"),
        filename = NULL,
        output=TRUE,
        
        # Circles
        lwd = 1,
        lty = 'blank',
        fill = myCol[1:2],
        
        # Numbers
        cex = .6,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 0.6,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.pos = c(0, 0)
)

t_cells_vienn <- venn.diagram(
        x = list(deseq_res_study$fetahu$t_cell$res %>% 
                   dplyr::filter(padj < 0.05) %>% .$gene,
                 deseq_res_study$dong$t_cell$res %>% 
                   dplyr::filter(padj < 0.05) %>% .$gene,
                 deseq_res_study$verhoeven$b_cell$res %>% 
                   dplyr::filter(padj < 0.05) %>% .$gene),
        category.names = c("Fetahu" , "Dong", "Verhoeven"),
        filename = NULL,
        output=TRUE,
        
        # Circles
        lwd = 1,
        lty = 'blank',
        fill = myCol[1:3],
        
        # Numbers
        cex = .6,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 0.6,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.pos = c(0, 0, 180)
)

m_cells_vienn <- venn.diagram(
        x = list(deseq_res_study$fetahu$myeloid_cell$res %>% 
                   dplyr::filter(padj < 0.05) %>% .$gene,
                 deseq_res_study$dong$myeloid_cell$res %>% 
                   dplyr::filter(padj < 0.05) %>% .$gene,
                 deseq_res_study$verhoeven$myeloid_cell$res %>% 
                   dplyr::filter(padj < 0.05) %>% .$gene,
                 deseq_res_study$jansky$myeloid_cell$res %>% 
                   dplyr::filter(padj < 0.05) %>% .$gene,
                 deseq_res_study$yuan$myeloid_cell$res %>% 
                   dplyr::filter(padj < 0.05) %>% .$gene),
        category.names = c("Fetahu" , "Dong", "Verhoeven", 'Jansky', 'Yuan'),
        filename = NULL,
        output=TRUE,
        
        # Circles
        lwd = 1,
        lty = 'blank',
        fill = myCol,
        
        # Numbers
        cex = .6,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 0.6,
        cat.fontface = "bold",
        cat.default.pos = "outer"
)

plot_grid(b_cells_vienn, t_cells_vienn, m_cells_vienn,
          nrow = 1, 
          labels = c('B-cells', 'T-cells', 'Myeloid cells'),
          rel_widths = c(1.1, 1.3, 1.8))
```

### Scatter plots
```{r}
merged_df <- merge(deseq_res_study$dong$t_cell$res, 
                 deseq_res_study$verhoeven$t_cell$res,
                 by = "gene", suffixes = c("_dataset1", "_dataset2"))

# Create scatter plot
ggplot(merged_df, aes(x = log2FoldChange_dataset1, y = log2FoldChange_dataset2)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "red") +
  labs(x = "Log2 Fold Change (Fetahu)", y = "Log2 Fold Change (Verhoeven)",
       title = "Scatter Plot of Log2 Fold Changes") +
  theme_minimal()
```


```{r eval=FALSE, include=FALSE}
limma_correct_batch <- function(list_matrix, coldata, cell_type, design) {
  
  count_matrix <- as.matrix(list_matrix[cell_type][[1]])
  coldata_cell <- coldata %>% dplyr::filter(ANNOTATION_FINAL == cell_type) %>%
                                set_rownames(rownames(.) %>% str_extract("[A-Za-z0-9-]+_(.*)",
                                                                         group = 1))
  
  group <- str_extract(colnames(count_matrix), "[A-Za-z0-9-]+_(.*)_", group = 1)  
  dge <- DGEList(counts = count_matrix, samples = coldata_cell, 
                 group = group)
  
  # Normalize counts
  dge <- calcNormFactors(dge)
  
  # Design matrix
  batch <- str_extract(colnames(count_matrix), "[A-Za-z0-9-]+-(.*)_(.*)_", group = 1)  
  design <- model.matrix(~ batch + group)
  
  # Estimate dispersion
  dge <- estimateDisp(dge, design)
  
  # Fit the model
  fit <- glmFit(dge, design)
  
  # Remove batch effects
  logCPM <- cpm(dge, log = TRUE)
  logCPM_batch_corrected <- removeBatchEffect(logCPM, batch = batch)
  
  # Perform differential expression analysis
  contrast <- makeContrasts(group2 - group1, levels = design)
  lrt <- glmLRT(fit, contrast = contrast)
  topTags(lrt)
}

```

```{r eval=FALSE, include=FALSE}

limma_correct_batch <- function(list_matrix, coldata, cell_type, design) {
  
  count_matrix <- as.matrix(list_matrix[cell_type][[1]])
  coldata_cell <- coldata %>% dplyr::filter(ANNOTATION_FINAL == cell_type) %>%
                                set_rownames(rownames(.) %>% str_extract("[A-Za-z0-9-]+_(.*)",
                                                                         group = 1))
  dds <- DESeqDataSetFromMatrix(count_matrix, 
                              colData = coldata_cell, 
                              design = design)
  rld <- vst(dds, blind=TRUE)
  
  DESeq2::plotPCA(rld, intgroup=c("less18M", "Study"))
  
  lcpm <- cpm(count_matrix, log=TRUE)
  
  batch <- str_extract(colnames(count_matrix), "[A-Za-z0-9-]+-(.*)_(.*)_", group = 1)  
  lcpm_corrected <- removeBatchEffect(lcpm, batch = batch)
  
  group <- str_extract(colnames(count_matrix), "[A-Za-z0-9-]+_(.*)_", group = 1)  %>%  as.factor()
  limma::plotMDS(lcpm_corrected, col = as.numeric(group))
}


```

# Find similar genes

```{r}
merged_t_cells <- merge(deseq_res_study$dong$t_cell$res %>% dplyr::select(gene, log2FoldChange, pvalue, padj, lfcSE), 
                 deseq_res_study$verhoeven$t_cell$res %>% dplyr::select(gene, log2FoldChange, pvalue, padj, lfcSE), 
                 
                 by = "gene", suffixes = c("_dong", "_verhoeven")) %>% 
  merge(deseq_res_study$fetahu$t_cell$res %>% dplyr::select(gene, log2FoldChange, pvalue, padj, lfcSE) %>% 
          set_colnames(paste0(colnames(.), '_fetahu')) %>% mutate(gene = gene_fetahu) %>% dplyr::select(-gene_fetahu), 
        by = "gene")

merged_b_cells <- merge(deseq_res_study$fetahu$b_cell$res %>% dplyr::select(gene, log2FoldChange, pvalue, padj, lfcSE), 
                 deseq_res_study$verhoeven$b_cell$res %>% dplyr::select(gene, log2FoldChange, pvalue, padj, lfcSE), 
                 
                 by = "gene", suffixes = c("_fetahu", "_verhoeven"))

merged_myeloid_cells <- merge(deseq_res_study$dong$myeloid_cell$res %>% dplyr::select(gene, log2FoldChange, pvalue, padj, lfcSE), 
                 deseq_res_study$verhoeven$myeloid_cell$res %>% dplyr::select(gene, log2FoldChange, pvalue, padj, lfcSE), 
                 
                 by = "gene", suffixes = c("_dong", "_verhoeven")) %>% 
  merge(deseq_res_study$fetahu$myeloid_cell$res %>% dplyr::select(gene, log2FoldChange, pvalue, padj, lfcSE) %>% 
          set_colnames(paste0(colnames(.), '_fetahu')) %>% mutate(gene = gene_fetahu) %>% dplyr::select(-gene_fetahu), 
        by = "gene") %>%
  merge(deseq_res_study$yuan$myeloid_cell$res %>% dplyr::select(gene, log2FoldChange, pvalue, padj, lfcSE) %>% 
          set_colnames(paste0(colnames(.), '_yuan')) %>% mutate(gene = gene_yuan) %>% dplyr::select(-gene_yuan), 
        by = "gene") # %>% 
  # merge(deseq_res_study$jansky$myeloid_cell$res %>% dplyr::select(gene, log2FoldChange, pvalue, padj) %>% 
  #         set_colnames(paste0(colnames(.), '_jansky')) %>% mutate(gene = gene_jansky) %>% dplyr::select(-gene_jansky), 
  #       by = "gene")
```

```{r}
myeloid_genes_up <- merged_myeloid_cells %>%
  filter(across(all_of(grep("^log2FoldChange", names(.), value = TRUE)), ~ . > 0)) %>% .$gene
myeloid_genes_down <- merged_myeloid_cells %>%
  filter(across(all_of(grep("^log2FoldChange", names(.), value = TRUE)), ~ . < 0)) %>% .$gene

tcells_genes_up <- merged_t_cells %>%
  filter(across(all_of(grep("^log2FoldChange", names(.), value = TRUE)), ~ . > 0)) %>% .$gene
tcells_genes_down <- merged_t_cells %>%
  filter(across(all_of(grep("^log2FoldChange", names(.), value = TRUE)), ~ . < 0)) %>% .$gene

bcells_genes_up <- merged_b_cells %>%
  filter(across(all_of(grep("^log2FoldChange", names(.), value = TRUE)), ~ . > 0)) %>% .$gene
bcells_genes_down <- merged_b_cells %>%
  filter(across(all_of(grep("^log2FoldChange", names(.), value = TRUE)), ~ . < 0)) %>% .$gene
```

# Metaseq

i don't like this method

```{r eval=FALSE, include=FALSE}
metaseq_run <- function(count_data) {
  datasets = str_extract(colnames(count_data), "[A-Za-z0-9-]+-(.*)_(.*)_", group = 1) %>% unique()
  labels = c("TRUE", "FALSE")
  col_names <- colnames(count_data)
  # Function to check if a dataset has both TRUE and FALSE labels
  has_both_labels <- function(dataset, col_names) {
    has_true <- any(grepl(paste0(dataset, "_TRUE"), col_names))
    has_false <- any(grepl(paste0(dataset, "_FALSE"), col_names))
    has_true && has_false
  }
  # Identify datasets with both labels
  valid_datasets <- datasets[sapply(datasets, has_both_labels, col_names)]
  subset_cols <- unlist(lapply(valid_datasets, function(dataset) col_names[grepl(dataset, col_names)]))
  # Subset the matrix
  count_data <- count_data[, subset_cols, drop = FALSE]
  
  flag1 <- str_extract(colnames(count_data), "[A-Za-z0-9-]+_(.*)_", group = 1) # condition : true/false
  flag2 <- str_extract(colnames(count_data), "[A-Za-z0-9-]+-(.*)_(.*)_", group = 1) # study
  cds <- meta.readData(data = count_data, factor = flag1, studies = flag2)
  
  cl <- makeCluster(4, "SOCK")
  result <- meta.oneside.noiseq(cds, k = 0.5, norm = "tmm", replicates = "biological",
  factor = flag1, conditions = c('FALSE', 'TRUE'), studies = flag2, cl = cl)
  stopCluster(cl)
  
  return(result)
}

metaseq_t_cells_result <- metaseq_run(cts.split.modified$`T-cells`)
metaseq_t_cells_result <- metaseq_run(cts.split.modified$`T-cells`)
metaseq_myeloid_cells_result <- metaseq_run(cts.split.modified$`B-cells`)
```

```{r eval=FALSE, include=FALSE}
metaseq_b_cells_result_ft <- Fisher.test(metaseq_b_cells_result)
metaseq_b_cells_result_st <- Stouffer.test(metaseq_b_cells_result)
metaseq_t_cells_result_ft <- Fisher.test(metaseq_t_cells_result)
metaseq_t_cells_result_st <- Stouffer.test(metaseq_t_cells_result)
metaseq_myeloid_cells_result_ft <- Fisher.test(metaseq_myeloid_cells_result)
metaseq_myeloid_cells_result_st <- Stouffer.test(metaseq_myeloid_cells_result)

(metaseq_t_cells_result_ft$Lower < 0.05) %>% sum(., na.rm = TRUE)

```

let's try to combine p-values from deseq

# Combine pvalue - scran

```{r}
myeloid_genes_up <- merged_myeloid_cells %>%
  filter(across(all_of(grep("^log2FoldChange", names(.), value = TRUE)), ~ . > 0)) 
myeloid_genes_down <- merged_myeloid_cells %>%
  filter(across(all_of(grep("^log2FoldChange", names(.), value = TRUE)), ~ . < 0)) 

tcells_genes_up <- merged_t_cells %>%
  filter(across(all_of(grep("^log2FoldChange", names(.), value = TRUE)), ~ . > 0))
tcells_genes_down <- merged_t_cells %>%
  filter(across(all_of(grep("^log2FoldChange", names(.), value = TRUE)), ~ . < 0)) 

bcells_genes_up <- merged_b_cells %>%
  filter(across(all_of(grep("^log2FoldChange", names(.), value = TRUE)), ~ . > 0)) 
bcells_genes_down <- merged_b_cells %>%
  filter(across(all_of(grep("^log2FoldChange", names(.), value = TRUE)), ~ . < 0)) 
```

```{r}
t_cells_up_comp_pval <- combineParallelPValues(list(tcells_genes_up$pvalue_dong, 
                                    tcells_genes_up$pvalue_verhoeven, 
                                    tcells_genes_up$pvalue_fetahu),
                                    method='berger')
tcells_genes_up$combined_pval <- t_cells_up_comp_pval$p.value
t_cells_down_comp_pval <- combineParallelPValues(list(tcells_genes_down$pvalue_dong, 
                                    tcells_genes_down$pvalue_verhoeven, 
                                    tcells_genes_down$pvalue_fetahu),
                                    method='berger')
tcells_genes_down$combined_pval <- t_cells_down_comp_pval$p.value

b_cells_up_comp_pval <- combineParallelPValues(list(bcells_genes_up$pvalue_fetahu, 
                                    bcells_genes_up$pvalue_verhoeven),
                                    method='berger')
bcells_genes_up$combined_pval <- b_cells_up_comp_pval$p.value
b_cells_down_comp_pval <- combineParallelPValues(list(bcells_genes_down$pvalue_fetahu, 
                                    bcells_genes_down$pvalue_verhoeven),
                                    method='berger')
bcells_genes_down$combined_pval <- b_cells_down_comp_pval$p.value

m_cells_up_comp_pval <- combineParallelPValues(list(myeloid_genes_up$pvalue_dong, 
                                    myeloid_genes_up$pvalue_verhoeven, 
                                    myeloid_genes_up$pvalue_fetahu,
                                    myeloid_genes_up$pvalue_yuan),
                                    method='berger')
myeloid_genes_up$combined_pval <- m_cells_up_comp_pval$p.value
m_cells_down_comp_pval <- combineParallelPValues(list(myeloid_genes_down$pvalue_dong, 
                                    myeloid_genes_down$pvalue_verhoeven, 
                                    myeloid_genes_down$pvalue_fetahu,
                                    myeloid_genes_down$pvalue_yuan),
                                    method='berger')
myeloid_genes_down$combined_pval <- m_cells_down_comp_pval$p.value

tcells_genes_up$avg_log2FoldChange <- rowMeans(tcells_genes_up[, grep("log2FoldChange", names(tcells_genes_up), value = TRUE)], na.rm = TRUE)
tcells_genes_down$avg_log2FoldChange <- rowMeans(tcells_genes_down[, grep("log2FoldChange", names(tcells_genes_down), value = TRUE)], na.rm = TRUE)

bcells_genes_up$avg_log2FoldChange <- rowMeans(bcells_genes_up[, grep("log2FoldChange", names(bcells_genes_up), value = TRUE)], na.rm = TRUE)
bcells_genes_down$avg_log2FoldChange <- rowMeans(bcells_genes_down[, grep("log2FoldChange", names(bcells_genes_down), value = TRUE)], na.rm = TRUE)

myeloid_genes_up$avg_log2FoldChange <- rowMeans(myeloid_genes_up[, grep("log2FoldChange", names(myeloid_genes_up), value = TRUE)], na.rm = TRUE)
myeloid_genes_down$avg_log2FoldChange <- rowMeans(myeloid_genes_down[, grep("log2FoldChange", names(myeloid_genes_down), value = TRUE)], na.rm = TRUE)
```

```{r fig.height=8, fig.width=15}
volcano_plot_2df <- function(df1, df2) {
  
  df = rbind(df1, df2)
  
  EnhancedVolcano(df,
    lab = df$gene,
    x = 'avg_log2FoldChange',
    y = 'combined_pval',
    pCutoff = 0.05,
    FCcutoff=0.5) 
}


plot_grid(volcano_plot_2df(tcells_genes_up, tcells_genes_down) +
            ggtitle('T-cells'),
          volcano_plot_2df(bcells_genes_up, bcells_genes_down) +
            ggtitle('B-cells'),
          volcano_plot_2df(myeloid_genes_up, myeloid_genes_down) +
            ggtitle('Myeloid cells'), 
          nrow = 1)
```

# Forest plot

## Myeloid cells

```{r fig.height=8}
# Select the top 5 genes by avg_log2FoldChange
top_genes <- myeloid_genes_up[order(myeloid_genes_up$combined_pval), ][1:7, ] %>% 
  rbind(myeloid_genes_down[order(myeloid_genes_down$combined_pval), ][1:7, ]) 

top_genes <- top_genes[order(-top_genes$avg_log2FoldChange), ]

# Create a factor for gene to ensure the correct order in the plot
top_genes$gene <- factor(top_genes$gene, levels = top_genes$gene)

# Calculate 95% confidence intervals
top_genes$CI_lower_dong <- top_genes$log2FoldChange_dong - 1.96 * top_genes$lfcSE_dong
top_genes$CI_upper_dong <- top_genes$log2FoldChange_dong + 1.96 * top_genes$lfcSE_dong
top_genes$CI_lower_verhoeven <- top_genes$log2FoldChange_verhoeven - 1.96 * top_genes$lfcSE_verhoeven
top_genes$CI_upper_verhoeven <- top_genes$log2FoldChange_verhoeven + 1.96 * top_genes$lfcSE_verhoeven
top_genes$CI_lower_fetahu <- top_genes$log2FoldChange_fetahu - 1.96 * top_genes$lfcSE_fetahu
top_genes$CI_upper_fetahu <- top_genes$log2FoldChange_fetahu + 1.96 * top_genes$lfcSE_fetahu
top_genes$CI_lower_yuan <- top_genes$log2FoldChange_yuan - 1.96 * top_genes$lfcSE_yuan
top_genes$CI_upper_yuan <- top_genes$log2FoldChange_yuan + 1.96 * top_genes$lfcSE_yuan

# Prepare data in long format for ggplot2
library(tidyr)
library(dplyr)

forest_data <- top_genes %>%
  dplyr::select(gene, 
         starts_with("log2FoldChange"), 
         starts_with("CI_lower"), 
         starts_with("CI_upper")) %>%
  pivot_longer(cols = -gene, 
               names_to = c(".value", "study"), 
               names_pattern = "(log2FoldChange|CI_lower|CI_upper)_(.*)")

forest_data <- forest_data %>%
  mutate(study = factor(study, levels = c("dong", "verhoeven", "fetahu", "yuan")))

library(ggplot2)

ggplot(forest_data, aes(x = log2FoldChange, y = study, color = study)) +
  geom_point(position = position_dodge(width = 0.8)) +
  geom_errorbar(aes(xmin = CI_lower, xmax = CI_upper), width = 0.2, position = position_dodge(width = 0.8)) +
  scale_color_manual(values = c("dong" = "blue", "verhoeven" = "green", "fetahu" = "red", "yuan" = "purple")) +
  labs(x = "log2FoldChange", y = "Gene", title = "Myeloid cells") +
  theme_minimal() +
  theme(legend.title = element_blank(), 
  strip.text.y = element_text(angle=0)) +
  facet_grid(gene ~ .)
```

## T cells

```{r fig.height=7}
# Select the top 5 genes by avg_log2FoldChange
top_genes <- tcells_genes_up[order(tcells_genes_up$combined_pval), ][1:7, ] %>% 
  rbind(tcells_genes_down[order(tcells_genes_down$combined_pval), ][1:7, ]) 

top_genes <- top_genes[order(-top_genes$avg_log2FoldChange), ]

# Create a factor for gene to ensure the correct order in the plot
top_genes$gene <- factor(top_genes$gene, levels = top_genes$gene)

# Calculate 95% confidence intervals
top_genes$CI_lower_dong <- top_genes$log2FoldChange_dong - 1.96 * top_genes$lfcSE_dong
top_genes$CI_upper_dong <- top_genes$log2FoldChange_dong + 1.96 * top_genes$lfcSE_dong
top_genes$CI_lower_verhoeven <- top_genes$log2FoldChange_verhoeven - 1.96 * top_genes$lfcSE_verhoeven
top_genes$CI_upper_verhoeven <- top_genes$log2FoldChange_verhoeven + 1.96 * top_genes$lfcSE_verhoeven
top_genes$CI_lower_fetahu <- top_genes$log2FoldChange_fetahu - 1.96 * top_genes$lfcSE_fetahu
top_genes$CI_upper_fetahu <- top_genes$log2FoldChange_fetahu + 1.96 * top_genes$lfcSE_fetahu

# Prepare data in long format for ggplot2
library(tidyr)
library(dplyr)

forest_data <- top_genes %>%
  dplyr::select(gene, 
         starts_with("log2FoldChange"), 
         starts_with("CI_lower"), 
         starts_with("CI_upper")) %>%
  pivot_longer(cols = -gene, 
               names_to = c(".value", "study"), 
               names_pattern = "(log2FoldChange|CI_lower|CI_upper)_(.*)")

forest_data <- forest_data %>%
  mutate(study = factor(study, levels = c("dong", "verhoeven", "fetahu")))

library(ggplot2)

ggplot(forest_data, aes(x = log2FoldChange, y = study, color = study)) +
  geom_point(position = position_dodge(width = 0.8)) +
  geom_errorbar(aes(xmin = CI_lower, xmax = CI_upper), width = 0.2, position = position_dodge(width = 0.8)) +
  scale_color_manual(values = c("dong" = "blue", "verhoeven" = "green", "fetahu" = "red")) +
  labs(x = "log2FoldChange", y = "Gene", title = "T cells") +
  theme_minimal() +
  theme(legend.title = element_blank(), 
  strip.text.y = element_text(angle=0)) +
  facet_grid(gene ~ .)
```

## B cells

```{r}
# Select the top 5 genes by avg_log2FoldChange
top_genes <- bcells_genes_up[order(bcells_genes_up$combined_pval), ][1:7, ] %>% 
  rbind(bcells_genes_down[order(bcells_genes_down$combined_pval), ][1:7, ])

top_genes <- top_genes[order(-top_genes$avg_log2FoldChange), ]

# Create a factor for gene to ensure the correct order in the plot
top_genes$gene <- factor(top_genes$gene, levels = top_genes$gene)

# Calculate 95% confidence intervals
top_genes$CI_lower_verhoeven <- top_genes$log2FoldChange_verhoeven - 1.96 * top_genes$lfcSE_verhoeven
top_genes$CI_upper_verhoeven <- top_genes$log2FoldChange_verhoeven + 1.96 * top_genes$lfcSE_verhoeven
top_genes$CI_lower_fetahu <- top_genes$log2FoldChange_fetahu - 1.96 * top_genes$lfcSE_fetahu
top_genes$CI_upper_fetahu <- top_genes$log2FoldChange_fetahu + 1.96 * top_genes$lfcSE_fetahu

# Prepare data in long format for ggplot2
library(tidyr)
library(dplyr)

forest_data <- top_genes %>%
  dplyr::select(gene, 
         starts_with("log2FoldChange"), 
         starts_with("CI_lower"), 
         starts_with("CI_upper")) %>%
  pivot_longer(cols = -gene, 
               names_to = c(".value", "study"), 
               names_pattern = "(log2FoldChange|CI_lower|CI_upper)_(.*)")

forest_data <- forest_data %>%
  mutate(study = factor(study, levels = c("verhoeven", "fetahu")))

library(ggplot2)

ggplot(forest_data, aes(x = log2FoldChange, y = study, color = study)) +
  geom_point(position = position_dodge(width = 0.8)) +
  geom_errorbar(aes(xmin = CI_lower, xmax = CI_upper), width = 0.2, position = position_dodge(width = 0.8)) +
  scale_color_manual(values = c("verhoeven" = "green", "fetahu" = "red")) +
  labs(x = "log2FoldChange", y = "Gene", title = "B cells") +
  theme_minimal() +
  theme(legend.title = element_blank(), 
  strip.text.y = element_text(angle=0)) +
  facet_grid(gene ~ .)
```
