---
title: "Mice_report"
author: "Kate Petrenko"
date: "2024-07-31"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = FALSE,
	message = FALSE,
	warning = FALSE
)
library(data.table)
library(Seurat)
library(scDblFinder)
library(DoubletFinder)
library(dplyr)
library(remotes)
# remotes::install_github("carmonalab/STACAS")
library(STACAS)
library(SingleR)
# install.packages("scGate")
library(scGate)
library(ggplot2)
library(BiocParallel)
library(harmony)
library(RColorBrewer)
library(stringr)
library(cowplot)
library(scuttle)
# install.packages("devtools")
# BiocManager::install("Homo.sapiens")
# BiocManager::install("jpmam1/scalop") #forked copy of a really good guy 
# devtools::install_github("jlaffy/infercna")
library(infercna)
library(infercnv)
library(parallel)
library(future)
library(clustree)
library(dittoSeq)
# install.packages("ggh4x")
library(ggh4x)
#BiocManager::install('EnhancedVolcano')
library(EnhancedVolcano)
library(clusterProfiler)
# BiocManager::install('org.Mm.eg.db')
library("org.Mm.eg.db")
library(ReactomePA)
library(msigdbr)
# install.packages('qs')
library(qs)
# devtools::install_github("ncborcherding/escape")
# devtools::install_github("rcastelo/GSVA")
library(escape)
```

```{r global variables, include=FALSE}
set.seed(100)
DATADIR = '~/neuroblastoma/data/'
WORKDIR = '~/neuroblastoma/'
PREPRDATADIR = '~/neuroblastoma/preprocessing_data/'

my_colors <- c("#9ACD32FF", "#4682B4FF",  "#DDA0DDFF", "#FFA07AFF", "#8FBC8BFF",
               "#40E0D0FF",  "#F0E68CFF", "#5F9EA0FF", "#D2B48CFF",  
               "#6495EDFF", "#FF69B4FF", "#BA55D3FF",  "#F08080FF", "#32CD32FF",  
               "#FFDAB9FF",  "#87CEEBFF")
```

```{r download references, message=FALSE, warning=FALSE, include=FALSE}
mouse_ref <- celldex::MouseRNAseqData()
mouse_ref_imm <- celldex::ImmGenData()
scGate_models_DB <- get_scGateDB()
```

```{r include=FALSE}
young <- Read10X_h5(paste0(DATADIR,
                           'mice_scRNA-seq_v1/20240625_YuvalShaked_Chen/CellRanger/Counting_2024A/P_2024A/filtered_feature_bc_matrix.h5'))
adult <- Read10X_h5(paste0(DATADIR,
                           'mice_scRNA-seq_v1/20240625_YuvalShaked_Chen/CellRanger/Counting_2024A/A_2024A/filtered_feature_bc_matrix.h5'))
```

# QC - Filtration

## Young - before

We found that some cells have low complexity (log10(nFeature_RNA) / log10(nCount_RNA)). Now, complexity combines these two ideas. If a cell has a lot of RNA (high nCount_RNA) but is only expressing a few different genes (low nFeature_RNA), it has low complexity. This means most of the RNA comes from just a few genes. On the other hand, a cell with high complexity expresses many different genes, even if it doesn’t have a huge amount of RNA.

Cells with low complexity might be highly specialized or stressed, focusing on a few specific functions (like producing a lot of one type of protein). Cells with low complexity might not represent healthy, biologically relevant cells. Instead, they could be damaged, stressed, or even dead cells, which often have reduced and non-specific gene expression. Filtering out these cells helps ensure that the data analyzed reflects true biological activity rather than artifacts.


```{r}
young <- CreateSeuratObject(young, min.cells = 10, min.features = 200)
young$age <- 'young'
young[["percent.mt"]] <- PercentageFeatureSet(young, pattern = "^mt-")
# young[["percent.ribo"]] <- PercentageFeatureSet(young, pattern = "^Rp[sl]")
young[["complexity"]] <- log10(young$nFeature_RNA) / log10(young$nCount_RNA)
VlnPlot(young, c("nFeature_RNA", "nCount_RNA", "percent.mt", "complexity"), ncol = 4)
```

We can see that in our case low-complexity cells are in one cluster. So it can be explained by highly specialized cell type. So I decided to annotate this group of cells. 

```{r}
young <- young %>% 
    NormalizeData() %>%
    FindVariableFeatures()%>%
    ScaleData() %>%
    RunPCA() %>% 
    RunUMAP(dims = 1:20)
  
FeaturePlot(young, 'complexity')
```

It's probably erythroid cells. It's known fact that erythroid cells have low complexity. 

```{r fig.height=7, fig.width=10}
FeaturePlot(young, c('Hbb-bs', 'Hbb-bt', 'Hba-a1', 'Hba-a2'))
```

I am not sure that we need these cells in our analysis so I filtered them out.

## Young - after

```{r}
young <- subset(young, subset = percent.mt < 15 & complexity > 0.7)
VlnPlot(young, c("nFeature_RNA", "nCount_RNA", "percent.mt", "complexity"), ncol = 4)
```

## Adult - before

```{r}
adult <- CreateSeuratObject(adult, min.cells = 10, min.features = 200)
adult$age <- 'adult'
adult[["percent.mt"]] <- PercentageFeatureSet(adult, pattern = "^mt-")
# adult[["percent.ribo"]] <- PercentageFeatureSet(adult, pattern = "^Rp[sl]")
adult[["complexity"]] <- log10(adult$nFeature_RNA) / log10(adult$nCount_RNA)
VlnPlot(adult, c("nFeature_RNA", "nCount_RNA", "percent.mt", "complexity"), ncol = 4)
```

## Adult - after

In cells from adult mice we don't have the low-complexity group of cells. So I just filtered all cells with complexity lower 0.8

```{r}
adult <- subset(adult, subset = percent.mt < 15 & complexity > 0.8)
VlnPlot(adult, c("nFeature_RNA", "nCount_RNA", "percent.mt", "complexity"), ncol = 4)
```

# Annotation

I used SingleR (reference-based approach) and scGate (marker-based approach) for a cells annotation. I combined them to get high evidence cell type. 

```{r include=FALSE}

annotate <- function(seu) {
  
  seu <- seu %>% 
    NormalizeData() %>%
    FindVariableFeatures()%>%
    ScaleData() %>%
    RunPCA() %>% 
    RunUMAP(dims = 1:20) %>%
    FindNeighbors(dims = 1:20) %>%
    FindClusters(resolution = 1.5)
  
  sce <- as.SingleCellExperiment(seu)
  colLabels(sce) <- seu$RNA_snn_res.1.5
  Idents(seu) <- 'RNA_snn_res.1.5'
  
  # #cluster-wise annotation
  pred_bped_main <- SingleR(test = sce, ref = mouse_ref, clusters=colLabels(sce),
                            labels = mouse_ref$label.main, BPPARAM=MulticoreParam(30))
  # # Create a named vector for RenameIdents
  singleR_labels <- pred_bped_main$pruned.labels
  names(singleR_labels) <- rownames(pred_bped_main)
  
  # Update cluster identities in Seurat object based on SingleR results
  seu <- RenameIdents(seu, singleR_labels)
  seu[['SingleR_labels']] <- Idents(seu)
  
  # define malignant cells
  malignant_scGate_model <- gating_model(name = "Malignant", signature = c("MYCN"))
  seu <- scGate(data = seu, model = malignant_scGate_model)
  barcode_malignant <- unique(c(colnames(seu)[seu$is.pure == 'Pure']))
  
  # define stromal cells
  stromal_scGate_model <- gating_model(name = "Stromal", 
                                       signature = c('Mmp2', 'Col1a1', 'Col1a2', 'Col5a1', 'Lum', 'Pdgfra'), level = 1)
  stromal_scGate_model <- gating_model(stromal_scGate_model, name = 'Epithelial', 
                                      signature = c('Cdh1', 'Flt1'), level = 1)
  seu <- scGate(seu, model = stromal_scGate_model)
  barcode_stromal <- colnames(seu)[seu$is.pure == 'Pure']
  
  #define t-cells
  seu <- scGate(seu, model = scGate_models_DB$mouse$generic$Tcell)
  barcode_T <- colnames(seu)[seu$is.pure == 'Pure']
  
  #define b-cells
  bcells_scGate_model <- gating_model(name = "PanBcell", signature = c('Cd79a'), level = 1)
  bcells_scGate_model <- gating_model(bcells_scGate_model, name = 'Bcell', 
                                      signature = c('Ms4a1', 'Bank1', 'Pax5', 'Cd19'), level = 2)
  bcells_scGate_model <- gating_model(bcells_scGate_model, name = 'Plasma_cell', 
                                      signature = c('Igkc', 'Ighg3', 'Ighg1', 'Igha', 'Cd19-'), level = 2)
  seu <- scGate(seu, model = bcells_scGate_model)
  barcode_B <- colnames(seu)[seu$is.pure == 'Pure']
  
  #define myeloid cells
  seu <- scGate(seu, model = scGate_models_DB$mouse$generic$Myeloid)
  barcode_myeloid <- colnames(seu)[seu$is.pure == 'Pure']
  
  seu <- scGate(seu, model = scGate_models_DB$mouse$generic$MoMacDC)
  barcode_monocyte <- colnames(seu)[seu$is.pure == 'Pure']
  
  # Assign final labels
  seu$final_labels <- "unknown"
  seu$final_labels[colnames(seu) %in% barcode_malignant] <- "malignant_cells"
  seu$final_labels[colnames(seu) %in% barcode_T] <- "T_cells"
  seu$final_labels[colnames(seu) %in% barcode_B] <- "B_cells"
  seu$final_labels[colnames(seu) %in% barcode_myeloid] <- "myeloid_cells"
  seu$final_labels[colnames(seu) %in% barcode_monocyte] <- "myeloid_cells"
  seu$final_labels[colnames(seu) %in% barcode_stromal] <- "stromal_cells"
  
  annot_comp <- c("Adipocytes" = 'weird',
                  "Neurons" = 'malignant_cells',
                  "Astrocytes" = 'weird',
                  "Endothelial cells" = 'stromal_cells',
                  "Erythrocytes" = 'weird',
                  "Fibroblasts" = 'malignant_cells',      
                  "Granulocytes" = 'myeloid_cells',
                  "Macrophages" = 'myeloid_cells',
                  "Microglia" = 'weird',
                  "Monocytes" = 'myeloid_cells',
                  "NK cells" = 'T_cells',
                  "Oligodendrocytes" = 'weird',
                  "T cells" = 'T_cells',
                  "Dendritic cells" = 'weird',
                  "Cardiomyocytes" = 'weird',
                  "Hepatocytes" = 'weird',
                  "B cells" = 'B_cells',
                  "Epithelial cells" = 'stromal_cells', 
                  'unknown' = 'unknown')
  
  pred_bped_main$pruned.labels[is.na(pred_bped_main$pruned.labels)] <- 'unknown'
  
  pred_bped_main$pruned.labels <- annot_comp[as.character(pred_bped_main$pruned.labels)]
  
  final_annotation_df <- data.frame(cluster = rownames(pred_bped_main),
                                    annotation = '-')
  # Compare SingleR and scGate
  for (i in rownames(pred_bped_main)) {
    
    if (table(seu$RNA_snn_res.1.5)[i] == 0) {
      next
    }
    
    cells_cluster <- seu[, colnames(seu) %in%  colnames(seu)[seu$RNA_snn_res.1.5 == i]]
    SingleR_title <- pred_bped_main$pruned.labels[rownames(pred_bped_main) == i]
    scGate_max <- names(which.max(table(cells_cluster$final_labels)))
    scGate_max2 <- names(table(cells_cluster$final_labels) %>% sort(decreasing = T))[2]
    scGate_max2 <- ifelse(is.na(scGate_max2), 'unknown', scGate_max2)
    
    if (scGate_max == SingleR_title & scGate_max != 'unknown') {
      final_annotation_df$annotation[final_annotation_df$cluster == i] <- scGate_max
    } else if (scGate_max2 == SingleR_title & scGate_max2 != 'unknown') {
      final_annotation_df$annotation[final_annotation_df$cluster == i] <- scGate_max2
    } else if (scGate_max == 'unknown' & SingleR_title != 'unknown') {
      final_annotation_df$annotation[final_annotation_df$cluster == i] <- SingleR_title
    } else if (scGate_max != 'unknown' & SingleR_title == 'unknown') {
      final_annotation_df$annotation[final_annotation_df$cluster == i] <- scGate_max
    } else {
      final_annotation_df$annotation[final_annotation_df$cluster == i] <- 'unresolved'
    }
    
  }
  
  final_annotation <- final_annotation_df$annotation
  names(final_annotation) <- final_annotation_df$cluster
  
  Idents(seu) <- 'RNA_snn_res.1.5'
  seu <- RenameIdents(seu, final_annotation)
  seu[['ANNOTATION_FINAL']] <- Idents(seu)
  
  return(seu)
  
}
```

```{r eval=FALSE, include=FALSE}
annotated_samples <- mclapply(list(young, adult),
                              annotate, mc.cores = 2)

names(annotated_samples) <- c('P', 'A')

seu <- merge(annotated_samples[[1]], annotated_samples[[2]], add.cell.ids = c("P", "A"))
```

```{r eval=FALSE, include=FALSE}
seu <- seu %>% 
  NormalizeData() %>% 
  FindVariableFeatures() %>% 
  ScaleData() %>% 
  RunPCA(verbose=FALSE)
  
# seu <- seu %>% 
#   IntegrateLayers(
#     object = ., method = CCAIntegration,
#     orig.reduction = "pca", new.reduction = "integrated.cca",
#     verbose = FALSE) %>% 
#   RunUMAP(reduction = "integrated.cca", dims = 1:20) 

seu <- seu %>% 
  RunHarmony(., 'age',
             lambda = 1, verbose = FALSE) %>% 
  RunUMAP(reduction = "harmony", dims = 1:20, verbose=F) 

seu <- JoinLayers(seu)
```

```{r eval=FALSE, fig.height=5, fig.width=12, include=FALSE}
seu <- seu %>% 
  FindNeighbors(dims = 1:10) %>%
  FindClusters(resolution = 0.1) 

Idents(object = seu) <- "RNA_snn_res.0.1"

new.cluster.ids <- c('0'='malignant_cells',
                     '1'='malignant_cells',
                     '2'='T_cells',
                     '3'='B_cells',
                     '4'='myeloid_cells',
                     '5'='stromal_cells',
                     '6'='malignant_cells',
                     '7'='stromal_cells',
                     '8'='stromal_cells'
)

seu <- RenameIdents(seu, new.cluster.ids)
seu$annotation <- Idents(seu)

# qsave(seu, paste0(PREPRDATADIR, 'seu_mouse.qs'))
```

```{r echo=FALSE, fig.height=5, fig.width=12}
seu <- qread(paste0(PREPRDATADIR, 'seu_mouse.qs'))

Idents(object = seu) <- "annotation"

plot_grid(DimPlot(seu, cols = c("#FF69B4FF", "#40E0D0FF", "#6495EDFF",  "#9ACD32FF", "#BA55D3FF")) +
  ggtitle("Annotation"), 
  DimPlot(seu, group.by = 'age', cols = c("#6495EDFF", "#FF69B4FF")) +
  ggtitle("Age"),
  nrow = 1, rel_widths = c(1, 0.9))
```

We can validate this approach by checking the neuroblastoma marker (Mycn) in our cells.

```{r fig.height=5, fig.width=6}
FeaturePlot(seu, 'Mycn')
```

We observe more malignant cells in adult mice and a higher proportion of immune cells in young mice. However, comparing B cells between age groups is challenging due to the low number of B cells in adult mice.

```{r eval=FALSE, include=FALSE}
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
seu <- CellCycleScoring(seu, s.features = s.genes, g2m.features = g2m.genes)
```

```{r fig.height=4, fig.width=13}
plot_grid(dittoBarPlot(seu, 'age', group.by = 'annotation', color.panel = my_colors),
          dittoBarPlot(seu, 'Phase', group.by = 'annotation', color.panel = my_colors),
          dittoBarPlot(seu, 'annotation', group.by = 'age', color.panel = my_colors),
          rel_widths = c(1, 1, 0.7),
          nrow = 1)
```

# DEG analysis based on single-cell level

Initially, I conducted DEG analysis at the single-cell level. While this approach can identify a greater number of DE genes, it provides less confidence in the reliability of these findings.

```{r fig.height=8, fig.width=20}
Idents(seu) <- "age"
t_cells.de.markers <- FindMarkers(subset(seu, annotation == 'T_cells'), 
                                  ident.1 = "young", ident.2 = "adult", 
                                  min.pct = 0.25)

b_cells.de.markers <- FindMarkers(subset(seu, annotation == 'B_cells'), 
                                  ident.1 = "young", ident.2 = "adult", 
                                  min.pct = 0.25)

myeloid.de.markers <- FindMarkers(subset(seu, annotation == 'myeloid_cells'), 
                                  ident.1 = "young", ident.2 = "adult", 
                                  min.pct = 0.25)

malignant.de.markers <- FindMarkers(subset(seu, annotation == 'malignant_cells'), 
                                  ident.1 = "young", ident.2 = "adult", 
                                  min.pct = 0.25)

volcano_plot <- function(de.markers) {
  EnhancedVolcano(de.markers,
    lab = rownames(de.markers),
    x = 'avg_log2FC',
    y = 'p_val_adj',
    col = c("grey30","#8FBC8BFF", "#6495EDFF", "#BA55D3FF"),
    drawConnectors = TRUE,
    pCutoff = 0.05, 
    FCcutoff = 0.5,
    max.overlaps = 30
    ) +
  labs(subtitle = "Young vs Adult")
}

plot_grid(volcano_plot(t_cells.de.markers) +
            ggtitle('T-cells'),
          volcano_plot(b_cells.de.markers) +
            ggtitle('B-cells'),
          volcano_plot(myeloid.de.markers) +
            ggtitle('Myeloid cells'),
          volcano_plot(malignant.de.markers) +
            ggtitle('Malignant cells'),
          nrow = 1)
```

# DEG analysis based on pseudobulk

So i decided to run DEG analysis based on pseudobulk level. But it requires peplicates that we. don't have. I created pseudo replicates dividing our cells to 3 random groups - 3 replicates.

## Scatter plots

The plots compare gene expression between adult and young mice across different cell types, showing a strong overall correlation. However, several differentially expressed genes exhibit distinct patterns between age groups. 

```{r echo=FALSE, fig.height=7, fig.width=7}
seu$celltype.age <- paste(seu$annotation, seu$age, sep = "_")
Idents(seu) <- "celltype.age"
seu_bulk <- AggregateExpression(seu, group.by = c("annotation", "age"), return.seurat = TRUE)

p1 <- CellScatter(seu_bulk, "T-cells_young", "T-cells_adult")
p1 <- CellScatter(seu_bulk, "T-cells_young", "T-cells_adult", highlight = p1$data %>% filter(abs(T.cells_adult - T.cells_young) > 0.5) %>% rownames())
p1 <- LabelPoints(plot = p1, points = p1$data %>% filter(abs(T.cells_adult - T.cells_young) > 0.5) %>% rownames(), repel = TRUE)

p2 <- CellScatter(seu_bulk, "B-cells_young", "B-cells_adult")
p2 <- CellScatter(seu_bulk, "B-cells_young", "B-cells_adult", highlight = p2$data %>% filter(abs(B.cells_adult - B.cells_young) > 0.5) %>% rownames())
p2 <- LabelPoints(plot = p2, points = p2$data %>% filter(abs(B.cells_adult - B.cells_young) > 0.5) %>% rownames(), repel = TRUE)

p3 <- CellScatter(seu_bulk, "myeloid-cells_young", "myeloid-cells_adult")
p3 <- CellScatter(seu_bulk, "myeloid-cells_young", "myeloid-cells_adult", highlight = p3$data %>% filter(abs(myeloid.cells_adult - myeloid.cells_young) > 0.5) %>% rownames())
p3 <- LabelPoints(plot = p3, points = p3$data %>% filter(abs(myeloid.cells_adult - myeloid.cells_young) > 0.5) %>% rownames(), repel = TRUE)

p4 <- CellScatter(seu_bulk, "malignant-cells_young", "malignant-cells_adult")
p4 <- CellScatter(seu_bulk, "malignant-cells_young", "malignant-cells_adult", highlight = p4$data %>% filter(abs(malignant.cells_adult - malignant.cells_young) > 0.5) %>% rownames())
p4 <- LabelPoints(plot = p4, points =  p4$data %>% filter(abs(malignant.cells_adult - malignant.cells_young) > 0.5) %>% rownames(), repel = TRUE)

p1 + p2 + p3 + p4
```

```{r create pseudoreplicates, include=FALSE}
seu$rep <- NA
create_pseudorep <- function(seu) {
  
  rep1 <- c()
  rep2 <- c()
  rep3 <- c()
  reps <- list(rep1, rep2, rep3)
  
  samples <- seu$age %>% unique
  cell_types <- seu$annotation %>% unique
  sample_annot_table <- table(seu$annotation, seu$age)
  for (sample in samples) {
    for (celltype in cell_types) {
      
      if (sample_annot_table[celltype, sample] >= 150) {
        
        cells <- WhichCells(seu, expression = annotation ==  celltype & age == sample)
        cells_split <- split(cells,  cut(seq_along(cells), 3, labels = FALSE))
        reps[[1]] <- c(reps[[1]], cells_split[1])
        reps[[2]] <- c(reps[[2]], cells_split[2])
        reps[[3]] <- c(reps[[3]], cells_split[3])
      }
      else if (sample_annot_table[celltype, sample] >= 90) {
        cells <- WhichCells(seu, expression = annotation ==  celltype & age == sample)
        cells_split <- split(cells,  cut(seq_along(cells), 2, labels = FALSE))
        reps[[1]] <- c(reps[[1]], cells_split[1])
        reps[[2]] <- c(reps[[2]], cells_split[2])
      }
    }
  }
  return(reps)
}

replicates_list <- create_pseudorep(seu)

for (i in seq_along(replicates_list)) {
  replicate_name <- paste0("rep", i)
  cells <- unlist(replicates_list[[i]])
  seu$rep[cells] <- replicate_name
}
```

```{r include=FALSE}
seu_bulk <- subset(seu, rep %in% c('rep1', 'rep2', 'rep3'))
seu_bulk <- AggregateExpression(seu_bulk, return.seurat = T, slot = "counts", assays = "RNA", 
                            group.by = c("annotation", "age", 'rep'))
```

```{r fig.height=8, fig.width=20}
Idents(seu_bulk) <- "age"
t_cells.de.markers <- FindMarkers(subset(seu_bulk, annotation == 'T-cells'), 
                                  ident.1 = "young", ident.2 = "adult", 
                                  slot = "counts",
                                  test.use = "DESeq2", verbose = F)

b_cells.de.markers <- FindMarkers(subset(seu_bulk, annotation == 'B-cells'), 
                                  ident.1 = "young", ident.2 = "adult", slot = "counts",
                                  test.use = "DESeq2", verbose = F, 
                                  min.cells.group = 2)

myeloid.de.markers <- FindMarkers(subset(seu_bulk, annotation == 'myeloid-cells'), 
                                  ident.1 = "young", ident.2 = "adult", slot = "counts",
                                  test.use = "DESeq2", verbose = F)

malignant.de.markers <- FindMarkers(subset(seu_bulk, annotation == 'malignant-cells'), 
                                  ident.1 = "young", ident.2 = "adult", slot = "counts",
                                  test.use = "DESeq2", verbose = F)

plot_grid(volcano_plot(t_cells.de.markers) +
            ggtitle('T-cells'),
          volcano_plot(b_cells.de.markers) +
            ggtitle('B-cells'),
          volcano_plot(myeloid.de.markers) +
            ggtitle('Myeloid cells'),
          volcano_plot(malignant.de.markers) +
            ggtitle('Malignant cells'),
          nrow = 1)
```

## Heatmaps

The heatmaps show clear differences in gene expression between adult and young mice within T cells and malignant cells. The distinct clustering patterns indicate age-specific gene expression profiles, highlighting the significant impact of age on these cell populations.

```{r fig.height=20, fig.width=12}
heatmap_tcells <- dittoHeatmap(subset(seu, annotation == 'T_cells'), t_cells.de.markers %>% 
                            filter(p_val_adj < 0.05 & abs(avg_log2FC) > 1) %>% 
                            dplyr::arrange(desc(avg_log2FC)) %>% 
                            rownames(),
             order.by = 'age', annot.by = 'age', silent=T,
             annot.colors = c("#87CEEBFF", "#DDA0DDFF"), 
             main = 'T cells', 
             breaks=seq(-2, 2, 0.2), heatmap.colors = colorRampPalette(c("#6495EDFF", "white", "#FF69B4FF"))(length(seq(-2, 2, 0.2))))

heatmap_bcells <- dittoHeatmap(subset(seu, annotation == 'B_cells'), b_cells.de.markers %>% 
                            filter(p_val_adj < 0.05 & abs(avg_log2FC) > 0.5) %>% 
                            dplyr::arrange(desc(avg_log2FC)) %>% 
                            rownames(),
             order.by = 'age', annot.by = 'age', silent=T,
             annot.colors = c("#87CEEBFF", "#DDA0DDFF"), 
             main = 'B cells', 
             breaks=seq(-2, 2, 0.2), heatmap.colors = colorRampPalette(c("#6495EDFF", "white", "#FF69B4FF"))(length(seq(-2, 2, 0.2))))

heatmap_myelcells <- dittoHeatmap(subset(seu, annotation == 'myeloid_cells'), myeloid.de.markers %>% 
                            filter(p_val_adj < 0.05 & abs(avg_log2FC) > 0.8) %>% 
                            dplyr::arrange(desc(avg_log2FC)) %>% 
                            rownames(),
             order.by = 'age', annot.by = 'age', silent=T,
             annot.colors = c("#87CEEBFF", "#DDA0DDFF"), 
             main = 'Myeloid cells', 
             breaks=seq(-2, 2, 0.2), heatmap.colors = colorRampPalette(c("#6495EDFF", "white", "#FF69B4FF"))(length(seq(-2, 2, 0.2))))

heatmap_maligcells <- dittoHeatmap(subset(seu, annotation == 'malignant_cells'), malignant.de.markers %>% 
                            filter(p_val_adj < 0.05 & abs(avg_log2FC) > 0.5) %>% 
                            dplyr::arrange(desc(avg_log2FC)) %>% 
                            rownames(),
             order.by = 'age', annot.by = 'age', silent=T,
             annot.colors = c("#87CEEBFF", "#DDA0DDFF"), 
             main = 'Malignant cells', 
             breaks=seq(-2, 2, 0.2), heatmap.colors = colorRampPalette(c("#6495EDFF", "white", "#FF69B4FF"))(length(seq(-2, 2, 0.2))))

plot_grid(heatmap_tcells[[4]], heatmap_maligcells[[4]], nrow=1)
```

In contrast to T cells and malignant cells, the heatmaps for B cells and myeloid cells do not show clear differences in gene expression between adult and young mice. The clustering patterns in these cell types are less distinct, suggesting that age may have a less pronounced effect on gene expression in B cells and myeloid cells. It requires more detailed analysis. 

```{r fig.height=12, fig.width=12}
plot_grid(heatmap_bcells[[4]], heatmap_myelcells[[4]], nrow=1)
```

# T-cells

```{r fig.height=12, fig.width=8}
volcano_plot(t_cells.de.markers) +
            ggtitle('T-cells') +
  ylim(c(0, 75))
```

## Positive 

The bar plot displays the results of enrichment analysis for positive markers identified in T cells during DEG analysis. The analysis reveals significant involvement in immune response pathways, particularly “Costimulation by the CD28 family” and “Generation of second messenger molecules.” These pathways are crucial for T cell activation and signaling, highlighting the importance of immune-related processes among the differentially expressed genes identified in this analysis.

```{r fig.height=10, fig.width=12}
# ego_tcells <- enrichGO(gene          = t_cells.de.markers %>% 
#                             filter(p_val_adj < 0.05 & avg_log2FC > 0.5) %>% 
#                             rownames() %>% 
#                             mapIds(org.Mm.eg.db, 
#                                    keys = .,
#                                    column = "ENTREZID",
#                                    keytype = "SYMBOL"),
#                       OrgDb         = org.Mm.eg.db,
#                       ont           = "MF",
#                       pAdjustMethod = "BH",
#                       pvalueCutoff  = 0.05,
#                       qvalueCutoff  = 0.05,
#                       readable      = TRUE)
# 
# barplot(ego_tcells, showCategory=20) +
#   ggtitle('GO:MF')
# 
# wp_tcells <- enrichWP(t_cells.de.markers %>% 
#                             filter(p_val_adj < 0.05 & avg_log2FC > 0.5) %>% 
#                             rownames() %>% 
#                             mapIds(org.Mm.eg.db, 
#                                    keys = .,
#                                    column = "ENTREZID",
#                                    keytype = "SYMBOL"), organism = "Mus musculus")
# 
# barplot(wp_tcells, showCategory=20) +
#   ggtitle('WikiPathway')

reactome_tcells <- enrichPathway(gene=t_cells.de.markers %>% 
                            filter(p_val_adj < 0.05 & avg_log2FC > 0.5) %>% 
                            rownames() %>% 
                            mapIds(org.Mm.eg.db, 
                                   keys = .,
                                   column = "ENTREZID",
                                   keytype = "SYMBOL") %>% as.vector(), 
                            pvalueCutoff = 0.05, readable=TRUE,
                            organism = 'mouse')

barplot(reactome_tcells, showCategory=10, font.size = 14) +
  scale_fill_gradient(low = "#FF69B4FF", high = "#6495EDFF") +
  ggtitle('Reactome') +
  theme(plot.title = element_text(size=22),
        legend.text = element_text(size=12),
         legend.title = element_text(size=14),
        legend.key.size = unit(1, 'cm'))

```

## Negative 

The enrichment analysis of negative markers in T cells reveals that the pathways are primarily related to RNA and protein processes, such as translation and RNA processing. Since these pathways are fundamental and ubiquitous, the results are less interesting from a biological perspective, as they don’t provide specific insights into immune-related functions or other specialized processes.

```{r  fig.height=10, fig.width=12}
# ego_tcells <- enrichGO(gene          = t_cells.de.markers %>% 
#                             filter(p_val_adj < 0.05 & avg_log2FC < 0.5) %>% 
#                             rownames() %>% 
#                             mapIds(org.Mm.eg.db, 
#                                    keys = .,
#                                    column = "ENTREZID",
#                                    keytype = "SYMBOL"),
#                       OrgDb         = org.Mm.eg.db,
#                       ont           = "MF",
#                       pAdjustMethod = "BH",
#                       pvalueCutoff  = 0.05,
#                       qvalueCutoff  = 0.05,
#                       readable      = TRUE)
# 
# barplot(ego_tcells, showCategory=20) +
#   ggtitle('GO:MF')
# 
# wp_tcells <- enrichWP(t_cells.de.markers %>% 
#                             filter(p_val_adj < 0.05 & avg_log2FC < 0.5) %>% 
#                             rownames() %>% 
#                             mapIds(org.Mm.eg.db, 
#                                    keys = .,
#                                    column = "ENTREZID",
#                                    keytype = "SYMBOL"), organism = "Mus musculus")
# 
# barplot(wp_tcells, showCategory=20) +
#   ggtitle('WikiPathway')

reactome_tcells <- enrichPathway(gene=t_cells.de.markers %>% 
                            filter(p_val_adj < 0.05 & avg_log2FC < 0.5) %>% 
                            rownames() %>% 
                            mapIds(org.Mm.eg.db, 
                                   keys = .,
                                   column = "ENTREZID",
                                   keytype = "SYMBOL") %>% as.vector(), 
                            pvalueCutoff = 0.05, readable=TRUE,
                            organism = 'mouse')

barplot(reactome_tcells, showCategory=10, font.size = 14) +
  scale_fill_gradient(low = "#FF69B4FF", high = "#6495EDFF") +
  ggtitle('Reactome') +
  theme(plot.title = element_text(size=22),
        legend.text = element_text(size=12),
         legend.title = element_text(size=14),
        legend.key.size = unit(1, 'cm'))
```

## Composition

We found 4 clusters of T-cells.

```{r eval=FALSE, fig.height=5, fig.width=12, include=FALSE}
t_cells <- subset(seu, annotation == 'T_cells')

t_cells <- t_cells %>% 
  NormalizeData() %>% 
  FindVariableFeatures() %>% 
  ScaleData() %>% 
  RunPCA(verbose=FALSE) %>% 
  RunHarmony(., 'age',
             lambda = 1, verbose = FALSE) %>% 
  RunUMAP(reduction = "harmony", dims = 1:20, verbose=F) %>% 
  FindNeighbors(dims = 1:20) %>%
  FindClusters(resolution = 0.2)

new.cluster.ids <- c('0'='CD4_cells',
                     '1'='CD8_cells',
                     '2'='NK_cells',
                     '3'='CD4_Treg',
                     '4'='doublets',
                     '5'='ILC'
)

t_cells <- RenameIdents(t_cells, new.cluster.ids)
t_cells$de_annotation <- Idents(t_cells)

t_cells <- subset(t_cells, de_annotation != 'doublets')
t_cells <- subset(t_cells, percent.mt < 8)

t_cells <- t_cells %>% 
  # NormalizeData() %>% 
  # FindVariableFeatures() %>% 
  # ScaleData() %>% 
  # RunPCA(verbose=FALSE) %>% 
  # RunHarmony(., 'age',
  #            lambda = 1, verbose = FALSE) %>% 
  # RunUMAP(reduction = "harmony", dims = 1:10, verbose=F) %>% 
  FindNeighbors(dims = 1:15) %>%
  FindClusters(resolution = seq(0.1, 1, 0.1))

t_cells <- subset(t_cells, RNA_snn_res.0.2 != 5)
t_cells <- subset(t_cells, RNA_snn_res.0.2 != 4)

# clustree(t_cells)
```

```{r echo=FALSE, fig.height=5, fig.width=12}
t_cells <- qread(paste0(PREPRDATADIR, 't_cells_mouse_ssGSEAs3.qs'))

Idents(t_cells) <- t_cells$RNA_snn_res.0.2
plot_grid(DimPlot(t_cells, group.by = 'RNA_snn_res.0.2', cols = my_colors) +
  ggtitle("Clusters"), 
  DimPlot(t_cells, group.by = 'age', cols = c("#6495EDFF", "#FF69B4FF")) +
  ggtitle("Age"),
  nrow = 1, rel_widths = c(0.95, 1))
```

```{r fig.height=6, fig.width=12}

tcells.clusters.markers <- FindAllMarkers(t_cells, only.pos = TRUE, min.pct = 0.25)
top5 <- tcells.clusters.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>% 
  group_by(cluster) %>%
  top_n(n = 7, wt = avg_log2FC)

DotPlot(t_cells, top5$gene, cols = c('lightgrey', "#BA55D3FF")) +
  RotatedAxis()
```


## DEG by age

```{r fig.height=6, fig.width=8}
Idents(t_cells) <- 'age'
VlnPlot(t_cells, c('Cd4', 'Cd8a', 'Cd8b1', 'Ctla4', 'Ncr1', 'Klrk1'), cols=c("#FF69B4FF", "#6495EDFF"), ncol = 3)
```

## Annotation

```{r fig.height=10, fig.width=10}
Idents(t_cells) <- t_cells$RNA_snn_res.0.2
VlnPlot(t_cells, c('Cd3e', 'Cd3d', 'Cd3g', 'Klrb1c', 'Ncr1', 'Klrk1', 'Cd4', 'Cd8a', 'Cd8b1',
                   'Tbx21', 'Cxcr3', 'Il2ra', 'Entpd1', 'Foxp3','Ctla4'))
```

```{r include=FALSE}
new.cluster.ids <- c('0'='CD4_Tcells',
                     '1'='CD8_Tcells',
                     '2'='NK_cells',
                     '3'='CD4_Tcells'
)

t_cells <- RenameIdents(t_cells, new.cluster.ids)
t_cells$de_annotation <- Idents(t_cells)

Idents(t_cells) <- 'RNA_snn_res.0.2'
new.cluster.ids <- c('0'='CD4+ T cells',
                     '1'='CD8+ T cells',
                     '2'='NK cells',
                     '3'='CD4+CTLA4+ T cells'
)
t_cells <- RenameIdents(t_cells, new.cluster.ids)
t_cells$Annotation <- Idents(t_cells)

DimPlot(t_cells, cols = my_colors)
```

```{r echo=FALSE, fig.height=5, fig.width=12}
plot_grid(DimPlot(t_cells, group.by = 'Annotation', cols = my_colors) +
  ggtitle("Annotation"), 
  DimPlot(t_cells, group.by = 'age', cols = c("#6495EDFF", "#FF69B4FF")) +
  ggtitle("Age"),
  nrow = 1, rel_widths = c(1, 0.85))
```

```{r fig.height=4, fig.width=13}
plot_grid(dittoBarPlot(t_cells, 'age', group.by = 'Annotation', color.panel = my_colors),
          dittoBarPlot(t_cells, 'Phase', group.by = 'Annotation', color.panel = my_colors),
          dittoBarPlot(t_cells, 'Annotation', group.by = 'age', color.panel = my_colors),
          rel_widths = c(1, 1, 0.7),
          nrow = 1)

# qsave(t_cells, paste0(PREPRDATADIR, 't_cells_mouse_v2.qs'))
```

## Scatter plots

The plots compare gene expression between adult and young mice across different cell types, showing a strong overall correlation. The difference in hemoglobin gene expression in CD4+ T cells can likely be attributed to the small number of cells.

```{r echo=FALSE, fig.height=5, fig.width=15}
t_cells$celltype.age <- paste(t_cells$de_annotation, t_cells$age, sep = "_")
Idents(t_cells) <- "celltype.age"
t_cells_bulk <- AggregateExpression(t_cells, group.by = c("de_annotation", "age"), return.seurat = TRUE)

p1 <- CellScatter(t_cells_bulk, "CD4-Tcells_young", "CD4-Tcells_adult")
p1 <- CellScatter(t_cells_bulk, "CD4-Tcells_young", "CD4-Tcells_adult", highlight = p1$data %>% filter(abs(`CD4.Tcells_young` - `CD4.Tcells_adult`) > 1) %>% rownames())
p1 <- LabelPoints(plot = p1, points = p1$data %>% filter(abs(`CD4.Tcells_young` - `CD4.Tcells_adult`) > 1) %>% rownames(), repel = TRUE)

p2 <- CellScatter(t_cells_bulk, "CD8-Tcells_young", "CD8-Tcells_adult")
p2 <- CellScatter(t_cells_bulk, "CD8-Tcells_young", "CD8-Tcells_adult", highlight = p2$data %>% filter(abs(`CD8.Tcells_young` - `CD8.Tcells_adult`) > 1) %>% rownames())
p2 <- LabelPoints(plot = p2, points = p2$data %>% filter(abs(`CD8.Tcells_young` - `CD8.Tcells_adult`) > 1) %>% rownames(), repel = TRUE)

p3 <- CellScatter(t_cells_bulk, "NK-cells_young", "NK-cells_adult")
p3 <- CellScatter(t_cells_bulk, "NK-cells_young", "NK-cells_adult", highlight = p3$data %>% filter(abs(NK.cells_young - NK.cells_adult) > 1) %>% rownames())
p3 <- LabelPoints(plot = p3, points = p3$data %>% filter(abs(NK.cells_young - NK.cells_adult) > 1) %>% rownames(), repel = TRUE)

p1 + p2 + p3 
```

## DE based on single-cell level

We can see that we have not enough cells in adult group to compare CD8+ and CD4+ T cells

```{r fig.height=8, fig.width=20}
nk_cells.de.markers <- FindMarkers(t_cells, 
                                  ident.1 = "NK_cells_young", ident.2 = "NK_cells_adult", 
                                  min.pct = 0.25)
cd8_cells.de.markers <- FindMarkers(t_cells, 
                                  ident.1 = "CD8_Tcells_young", ident.2 = "CD8_Tcells_adult", 
                                  min.pct = 0.25)
cd4_cells.de.markers <- FindMarkers(t_cells, 
                                  ident.1 = "CD4_Tcells_young", ident.2 = "CD4_Tcells_adult", 
                                  min.pct = 0.25)

plot_grid(volcano_plot(nk_cells.de.markers) +
            ggtitle('NK cells'),
          volcano_plot(cd8_cells.de.markers) +
            ggtitle('CD8+ T cells'),
          volcano_plot(cd4_cells.de.markers) +
            ggtitle('CD4+ T cells'),
          nrow = 1, rel_widths = c(1, 0.5, 1))
```

## Single-cell Gene Set Enrichment Analysis

### NK cells

```{r include=FALSE}
GS.hallmark <- getGeneSets(species = 'Mus musculus', library = "H")
GS.reactome <- getGeneSets(species = 'Mus musculus', library = "C2", subcategory = 'CP:REACTOME')
GS.tf <- getGeneSets(species = 'Mus musculus', library = "C3", subcategory = 'GTRD')
```

```{r include=FALSE}
t_cells <- performNormalization(t_cells, 
                                      assay = "hallmark.ssGSEA", 
                                      gene.sets = GS.hallmark, 
                                      scale.factor = t_cells$nFeature_RNA)
hallmark.markers <- FindMarkers(t_cells, 
                              assay = "hallmark.ssGSEA_normalized", 
                              ident.1 = "NK_cells_young", ident.2 = "NK_cells_adult",
                              min.pct = 0,
                              logfc.threshold = 0)

t_cells <- performNormalization(t_cells, 
                                      assay = "reactome.ssGSEA", 
                                      gene.sets = GS.reactome, 
                                      scale.factor = t_cells$nFeature_RNA)
reactome.markers <- FindMarkers(t_cells, 
                              assay = "reactome.ssGSEA_normalized", 
                              ident.1 = "NK_cells_young", ident.2 = "NK_cells_adult",
                              min.pct = 0,
                              logfc.threshold = 0)

t_cells <- performNormalization(t_cells, 
                                      assay = "tf.ssGSEA", 
                                      gene.sets = GS.tf, 
                                      scale.factor = t_cells$nFeature_RNA)
tf.markers <- FindMarkers(t_cells, 
                              assay = "tf.ssGSEA_normalized", 
                              ident.1 = "NK_cells_young", ident.2 = "NK_cells_adult",
                              min.pct = 0,
                              logfc.threshold = 0)
```

**TGF-Beta Signaling:** Enriched in NK cells from adult mice. TGF-beta signaling is known to suppress NK cell activity, which could suggest a more immunosuppressive environment in adult mice’s tumors.

**NF-kB Activation (TRAF6 mediated):** Enriched in adult mice, a pathway that is often associated with inflammatory responses and could suggest heightened activity in NK cells from adult tumors.

```{r fig.width=13}
heatmapEnrichment(subset(t_cells, de_annotation == 'NK_cells'), 
                  group.by = "age",
                  gene.set.use = hallmark.markers %>% 
                    dplyr::filter(p_val_adj < 0.05) %>% 
                    dplyr::top_n(10, wt = abs(avg_log2FC)) %>% 
                    rownames(.),
                  assay = "hallmark.ssGSEA_normalized") +
  scale_fill_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) +
  theme(legend.position="right", legend.direction = "vertical") +
  ggtitle('Hallmarks (NK cells)')


```

```{r fig.width=13}
heatmapEnrichment(subset(t_cells, de_annotation == 'NK_cells'), 
                  group.by = "age",
                  gene.set.use = reactome.markers %>% 
                    dplyr::filter(p_val_adj < 0.05) %>% 
                    dplyr::top_n(10, wt = abs(avg_log2FC)) %>% 
                    rownames(.),
                  assay = "reactome.ssGSEA_normalized") +
  scale_fill_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) +
  theme(legend.position="right", legend.direction = "vertical") +
  ggtitle('Reactome (NK cells)')


```

```{r fig.width=13}
heatmapEnrichment(subset(t_cells, de_annotation == 'NK_cells'), 
                  group.by = "age",
                  gene.set.use = tf.markers %>% 
                    dplyr::filter(p_val_adj < 0.05) %>% 
                    dplyr::top_n(10, wt = abs(avg_log2FC)) %>% 
                    rownames(.),
                  assay = "tf.ssGSEA_normalized") +
  scale_fill_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) +
  theme(legend.position="right", legend.direction = "vertical") +
  ggtitle('TFs (NK cells)')


```

### CD4+ T cells

```{r include=FALSE}
t_cells <- performNormalization(t_cells, 
                                      assay = "hallmark.ssGSEA", 
                                      gene.sets = GS.hallmark, 
                                      scale.factor = t_cells$nFeature_RNA)
hallmark.markers <- FindMarkers(t_cells, 
                              assay = "hallmark.ssGSEA_normalized", 
                              ident.1 = "CD4_Tcells_young", ident.2 = "CD4_Tcells_adult",
                              min.pct = 0,
                              logfc.threshold = 0)

t_cells <- performNormalization(t_cells, 
                                      assay = "reactome.ssGSEA", 
                                      gene.sets = GS.reactome, 
                                      scale.factor = t_cells$nFeature_RNA)
reactome.markers <- FindMarkers(t_cells, 
                              assay = "reactome.ssGSEA_normalized", 
                              ident.1 = "CD4_Tcells_young", ident.2 = "CD4_Tcells_adult",
                              min.pct = 0,
                              logfc.threshold = 0)

t_cells <- performNormalization(t_cells, 
                                      assay = "tf.ssGSEA", 
                                      gene.sets = GS.tf, 
                                      scale.factor = t_cells$nFeature_RNA)
tf.markers <- FindMarkers(t_cells, 
                              assay = "tf.ssGSEA_normalized", 
                              ident.1 = "CD4_Tcells_young", ident.2 = "CD4_Tcells_adult",
                              min.pct = 0,
                              logfc.threshold = 0)
```

```{r fig.width=13}
heatmapEnrichment(subset(t_cells, de_annotation == 'CD4_Tcells'), 
                  group.by = "age",
                  gene.set.use = hallmark.markers %>% 
                    dplyr::filter(p_val_adj < 0.05) %>% 
                    dplyr::top_n(10, wt = abs(avg_log2FC)) %>% 
                    rownames(.),
                  assay = "hallmark.ssGSEA_normalized") +
  scale_fill_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) +
  theme(legend.position="right", legend.direction = "vertical") +
  ggtitle('Hallmarks (CD4+ T cells)')


```

```{r fig.width=13}
heatmapEnrichment(subset(t_cells, de_annotation == 'CD4_Tcells'), 
                  group.by = "age",
                  gene.set.use = reactome.markers %>% 
                    dplyr::filter(p_val_adj < 0.05) %>% 
                    dplyr::top_n(10, wt = abs(avg_log2FC)) %>% 
                    rownames(.),
                  assay = "reactome.ssGSEA_normalized") +
  scale_fill_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) +
  theme(legend.position="right", legend.direction = "vertical") +
  ggtitle('Reactome (CD4+ T cells)')


```

```{r fig.width=13}
heatmapEnrichment(subset(t_cells, de_annotation == 'CD4_Tcells'), 
                  group.by = "age",
                  gene.set.use = tf.markers %>% 
                    dplyr::filter(p_val_adj < 0.05) %>% 
                    dplyr::top_n(10, wt = abs(avg_log2FC)) %>% 
                    rownames(.),
                  assay = "tf.ssGSEA_normalized") +
  scale_fill_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) +
  theme(legend.position="right", legend.direction = "vertical") +
  ggtitle('TFs (CD4+ T cells)')


```

# Malignant cells

```{r fig.height=12, fig.width=8}
volcano_plot(malignant.de.markers) +
            ggtitle('Malignant cells')
```

```{r eval=FALSE, fig.height=5, fig.width=12, message=FALSE, warning=FALSE, include=FALSE}
malig_cells <- subset(seu, annotation == 'malignant_cells')

malig_cells <- malig_cells %>% 
  NormalizeData() %>% 
  FindVariableFeatures() %>% 
  ScaleData() %>% 
  RunPCA(verbose=FALSE) %>% 
  RunHarmony(., 'age',
             lambda = 1, verbose = FALSE) %>% 
  RunUMAP(reduction = "harmony", dims = 1:20, verbose=F) %>% 
  FindNeighbors(dims = 1:20) %>%
  FindClusters(resolution = 0.1)

malig_cells <- subset(malig_cells, RNA_snn_res.0.1 != 2)


malig_cells <- malig_cells %>% 
  NormalizeData() %>% 
  FindVariableFeatures() %>% 
  ScaleData(vars.to.regress = c("S.Score", "G2M.Score")) %>% 
  RunPCA(verbose=FALSE) %>% 
  RunHarmony(., 'age',
             lambda = 1, verbose = FALSE) %>% 
  RunUMAP(reduction = "harmony", dims = 1:20, verbose=F) %>% 
  FindNeighbors(dims = 1:20) %>%
  FindClusters(resolution = seq(0.1, 1, 0.1))

# 
malig_cells <- subset(malig_cells, RNA_snn_res.0.1 != 3)

# clustree(malig_cells)
```

```{r fig.height=5, fig.width=12,}
malig_cells <- qread(paste0(PREPRDATADIR, 'malig_cells_mouse_gsvas3_ssGSEA.qs'))

Idents(malig_cells) <- malig_cells$RNA_snn_res.0.1
plot_grid(DimPlot(malig_cells, group.by = 'RNA_snn_res.0.1', cols = my_colors) +
  ggtitle("Clusters"), 
  DimPlot(malig_cells, group.by = 'age', cols = c("#6495EDFF", "#FF69B4FF")) +
  ggtitle("Age"),
  nrow = 1)
```

```{r fig.height=4, fig.width=13}
plot_grid(dittoBarPlot(malig_cells, 'age', group.by = 'RNA_snn_res.0.1', color.panel = my_colors),
          dittoBarPlot(malig_cells, 'Phase', group.by = 'RNA_snn_res.0.1', color.panel = my_colors),
          dittoBarPlot(malig_cells, 'RNA_snn_res.0.1', group.by = 'age', color.panel = my_colors),
          rel_widths = c(1, 1, 0.7),
          nrow = 1)
```

The markers do not distinctly separate the clusters, dividing the cells by these clusters might not be the best approach. So I analysed all cells together.

```{r fig.height=6, fig.width=12}

maligcells.clusters.markers <- FindAllMarkers(malig_cells, only.pos = TRUE, min.pct = 0.25)
top5 <- maligcells.clusters.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>% 
  group_by(cluster) %>%
  top_n(n = 7, wt = avg_log2FC)

DotPlot(malig_cells, top5$gene, cols = c('lightgrey', "#BA55D3FF")) +
  RotatedAxis()
```

Mycn expression is found in all clusters. However, only clusters 0 and 1 are explicitly annotated as neuroblastoma cells. Therefore, at the moment the analysis has been carried out only on them

```{r}
FeaturePlot(malig_cells, 'Mycn')
```

## Enrichment

```{r fig.height=10, fig.width=12}

ego_nkcells_pos <- enrichGO(gene = malignant.de.markers %>% 
                            filter(p_val_adj < 0.05 & avg_log2FC > 0.5) %>% 
                            rownames() %>% 
                            mapIds(org.Mm.eg.db, 
                                   keys = .,
                                   column = "ENTREZID",
                                   keytype = "SYMBOL"),
                      OrgDb         = org.Mm.eg.db,
                      ont           = "BP",
                      pAdjustMethod = "BH",
                      pvalueCutoff  = 0.05,
                      qvalueCutoff  = 0.05,
                      readable      = TRUE)

barplot(ego_nkcells_pos, showCategory=10, font.size = 14) +
  scale_fill_gradient(low = "#FF69B4FF", high = "#6495EDFF") +
  ggtitle('Malignant (positive)') +
  theme(plot.title = element_text(size=22),
        legend.text = element_text(size=12),
         legend.title = element_text(size=14),
        legend.key.size = unit(1, 'cm'))

ego_nkcells_neg <- enrichGO(gene = malignant.de.markers %>% 
                            filter(p_val_adj < 0.05 & avg_log2FC < -0.5) %>% 
                            rownames() %>% 
                            mapIds(org.Mm.eg.db, 
                                   keys = .,
                                   column = "ENTREZID",
                                   keytype = "SYMBOL"),
                      OrgDb         = org.Mm.eg.db,
                      ont           = "BP",
                      pAdjustMethod = "BH",
                      pvalueCutoff  = 0.05,
                      qvalueCutoff  = 0.05,
                      readable      = TRUE)

barplot(ego_nkcells_neg, showCategory=10, font.size = 14) +
  scale_fill_gradient(low = "#FF69B4FF", high = "#6495EDFF") +
  ggtitle('Malignant (negative)') +
  theme(plot.title = element_text(size=22),
        legend.text = element_text(size=12),
         legend.title = element_text(size=14),
        legend.key.size = unit(1, 'cm'))
```

## Some interesting genes

We have Fos that is up-regulated in young mice.  It relates to the differentiation of neuroblastoma to neuronal cells. Also, other Junb gene is up-regulated. They are in one AP-1 complex.
I found an article. It's really old. But they say: "The increase in the c-fos gene expression is observed during the differentiation of NB-1 cells into neuronal cells"

Nfkbia:

Deletion and low expression of NFKBIA are associated with poor prognosis in lower-grade glioma patients

Klf4:

KLF4 can act as a tumor suppressor in human malignant neuroblastoma, which is a deadly tumor mostly in children, by inhibiting the cell cycle and activating the cell differentiation and death pathways.

Btg2:

BTG anti-proliferation factor 2
 
```{r fig.height=8}
Idents(malig_cells) <- 'age'
VlnPlot(malig_cells, c('Fos', 'Junb', 'Nfkbia', 'Klf4', 'Btg2', 'Irf1'), cols=c("#FF69B4FF", "#6495EDFF"))
```

## Immediate Early genes

```{r fig.height=8, fig.width=8}
genes <- c(
  "Atf3", "Bhlhe40", "Ccl2", "Ccn1", "Ccn2", "Ccnl1", "Cebpd", 
  "Csrnp1", "Cxcl3", "Dusp1", "Dusp5", "Dusp6", "Egr1", 
  "Egr3", "F3", "Fos", "Fosb", "Gadd45b", "Myc",
  "Gem", "Hbegf", "Ier3", "Il6", "Jun", "Junb", "Klf10", 
  "Klf6", "Klhl21", "Ldlr", "Mcl1", "Nfkbia", "Nfkbiz", 
  "Nr2c2", "Nr4a1", "Nr4a2", "Plau", "Pmaip1", "Rcan1", 
  "Sgk1", "Slc2a3", "Srf", "Tnfaip3", "Trib1", "Tsc22d1", 
  "Zfp36"
)

heatmap_maligcells <- dittoHeatmap(malig_cells, genes,
             order.by = 'age', annot.by = 'age', 
             silent=T,
             annot.colors = c("#87CEEBFF", "#DDA0DDFF"), 
             main = 'Malignant cells', 
             breaks=seq(-2, 2, 0.2), heatmap.colors = colorRampPalette(c("#6495EDFF", "white", "#FF69B4FF"))(length(seq(-2, 2, 0.2))))

plot_grid(heatmap_maligcells[[4]], nrow=1)
```

## Single-cell Gene Set Enrichment Analysis

```{r eval=FALSE, include=FALSE}
malig_cells <- escape::runEscape(malig_cells, 
                        method = "ssGSEA",
                        gene.sets = GS.hallmark, 
                        groups = 500, 
                        min.size = 5,
                        new.assay.name = "hallmark.ssGSEA",
                        BPPARAM = MulticoreParam(15))

qsave(malig_cells, paste0(PREPRDATADIR, 'malig_cells_mouse_gsvas1.qs'))

malig_cells <- escape::runEscape(malig_cells, 
                                 method = "ssGSEA",
                                 gene.sets = GS.reactome,
                                 groups = 500,
                                 min.size = 5,
                                 new.assay.name = "reactome.ssGSEA",
                                 BPPARAM = MulticoreParam(20))
qsave(malig_cells, paste0(PREPRDATADIR, 'malig_cells_mouse_gsvas2.qs'))

malig_cells <- escape::runEscape(malig_cells, 
                        method = "ssGSEA",
                        gene.sets = GS.tf, 
                        groups = 6000, 
                        min.size = 5,
                        new.assay.name = "tf.ssGSEA",
                        BPPARAM = MulticoreParam(15))

qsave(malig_cells, paste0(PREPRDATADIR, 'malig_cells_mouse_gsvas3_ssGSEA.qs'))
```

```{r include=FALSE}
malig_cells <- performNormalization(malig_cells, 
                                      assay = "hallmark.ssGSEA", 
                                      gene.sets = GS.hallmark, 
                                      scale.factor = malig_cells$nFeature_RNA)
hallmark.markers <- FindMarkers(malig_cells, 
                              assay = "hallmark.ssGSEA_normalized", 
                              ident.1 = "young", ident.2 = "adult",
                              min.pct = 0,
                              logfc.threshold = 0)

malig_cells <- performNormalization(malig_cells, 
                                      assay = "reactome.ssGSEA", 
                                      gene.sets = GS.reactome, 
                                      scale.factor = malig_cells$nFeature_RNA)
reactome.markers <- FindMarkers(malig_cells, 
                              assay = "reactome.ssGSEA_normalized", 
                              ident.1 = "young", ident.2 = "adult",
                              min.pct = 0,
                              logfc.threshold = 0)

malig_cells <- performNormalization(malig_cells, 
                                      assay = "tf.ssGSEA", 
                                      gene.sets = GS.tf, 
                                      scale.factor = malig_cells$nFeature_RNA)
tf.markers <- FindMarkers(malig_cells, 
                              assay = "tf.ssGSEA_normalized", 
                              ident.1 = "young", ident.2 = "adult",
                              min.pct = 0,
                              logfc.threshold = 0)
```

```{r fig.width=13}
heatmapEnrichment(malig_cells, 
                  group.by = "ident",
                  gene.set.use = hallmark.markers %>% 
                    dplyr::filter(p_val_adj < 0.05) %>% 
                    dplyr::top_n(10, wt = abs(avg_log2FC)) %>% 
                    rownames(.),
                  assay = "hallmark.ssGSEA_normalized") +
  scale_fill_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) +
  theme(legend.position="right", legend.direction = "vertical") +
  ggtitle('Hallmarks')


```

```{r fig.width=13}
heatmapEnrichment(malig_cells, 
                  group.by = "ident",
                  gene.set.use = reactome.markers %>% 
                    dplyr::filter(p_val_adj < 0.05) %>% 
                    dplyr::top_n(10, wt = abs(avg_log2FC)) %>% 
                    rownames(.),
                  assay = "reactome.ssGSEA_normalized") +
  scale_fill_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) +
  theme(legend.position="right", legend.direction = "vertical") +
  ggtitle('Reactome')


```

```{r fig.width=13}
heatmapEnrichment(malig_cells, 
                  group.by = "ident",
                  gene.set.use = tf.markers %>% 
                    dplyr::filter(p_val_adj < 0.05) %>% 
                    dplyr::top_n(10, wt = abs(avg_log2FC)) %>% 
                    rownames(.),
                  assay = "tf.ssGSEA_normalized") +
  scale_fill_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) +
  theme(legend.position="right", legend.direction = "vertical") +
  ggtitle('TFs')


```

```{r eval=FALSE, include=FALSE}
geyserEnrichment(malig_cells, 
                 assay = "tf.ssGSEA_normalized",
                 gene.set = "ZNF454-TARGET-GENES")
```

```{r eval=FALSE, include=FALSE}
hallmark.markers %>% 
  dplyr::filter(p_val_adj < 0.05) %>% 
  dplyr::arrange(desc(abs(avg_log2FC)))

reactome.markers  %>% 
  dplyr::filter(p_val_adj < 0.05) %>% 
  dplyr::arrange(desc(abs(avg_log2FC)))

tf.markers  %>% 
  dplyr::filter(p_val_adj < 0.05) %>% 
  dplyr::arrange(desc(abs(avg_log2FC)))
```
