---
title: "Myeloid_cells_analysis"
output: html_document
date: "2024-08-24"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = FALSE,
	message = FALSE,
	warning = FALSE
)
library(data.table)
library(Seurat)
library(scDblFinder)
library(DoubletFinder)
library(dplyr)
library(remotes)
# remotes::install_github("carmonalab/STACAS")
library(STACAS)
library(SingleR)
# install.packages("scGate")
library(scGate)
library(ggplot2)
library(BiocParallel)
library(harmony)
library(RColorBrewer)
library(stringr)
library(cowplot)
library(scuttle)
# install.packages("devtools")
# BiocManager::install("Homo.sapiens")
# BiocManager::install("jpmam1/scalop") #forked copy of a really good guy 
# devtools::install_github("jlaffy/infercna")
library(infercna)
# library(infercnv)
library(parallel)
library(future)
library(clustree)
library(dittoSeq)
# install.packages("ggh4x")
library(ggh4x)
#BiocManager::install('EnhancedVolcano')
library(EnhancedVolcano)
library(clusterProfiler)
# BiocManager::install('org.Mm.eg.db')
library("org.Mm.eg.db")
library(ReactomePA)
library(msigdbr)
# install.packages('qs')
library(qs)
# devtools::install_github("ncborcherding/escape")
# devtools::install_github("rcastelo/GSVA")
library(escape)
# install.packages('gprofiler2')
library(gprofiler2)
# remotes::install_github("carmonalab/ProjecTILs")
library(ProjecTILs)
library(gridExtra)
```

```{r global variables server, eval=FALSE, include=FALSE}
set.seed(100)
DATADIR = '~/neuroblastoma/data/'
WORKDIR = '~/neuroblastoma/'
PREPRDATADIR = '~/neuroblastoma/preprocessing_data/'

my_colors <- c("#9ACD32FF", "#4682B4FF",  "#DDA0DDFF", "#FFA07AFF", "#8FBC8BFF",
               "#40E0D0FF",  "#F0E68CFF", "#5F9EA0FF", "#D2B48CFF",  
               "#6495EDFF", "#FF69B4FF", "#BA55D3FF",  "#F08080FF", "#32CD32FF",  
               "#FFDAB9FF",  "#87CEEBFF")
```

```{r download references, message=FALSE, warning=FALSE, include=FALSE}
mouse_ref <- celldex::MouseRNAseqData()
mouse_ref_imm <- celldex::ImmGenData()
scGate_models_DB <- get_scGateDB()
```

```{r include=FALSE}
seu <- qread(paste0(PREPRDATADIR, 'seu_mouse.qs'))
```

```{r include=FALSE}
myeloid <- subset(seu, annotation == 'myeloid_cells')
```

```{r include=FALSE}
mmus_s = gorth(cc.genes$s.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name
mmus_g2m = gorth(cc.genes$g2m.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name

myeloid <- CellCycleScoring(myeloid, s.features = mmus_s, g2m.features = mmus_g2m)
```

```{r include=FALSE}
myeloid <- myeloid %>% 
    NormalizeData() %>%
    FindVariableFeatures()%>%
    ScaleData() %>%
    RunPCA()

myeloid <- myeloid %>% 
    RunUMAP(dims = 1:4) %>%
    FindNeighbors(dims = 1:4) %>%
    FindClusters(resolution = seq(0.1, 1, 0.1))
```

```{r eval=FALSE, fig.height=20, fig.width=10, include=FALSE}
DimHeatmap(myeloid, dims = 1:20)
```

```{r eval=FALSE, include=FALSE}
ElbowPlot(myeloid)
```

```{r include=FALSE}
DimPlot(myeloid, group.by = 'RNA_snn_res.0.1')
DimPlot(myeloid, group.by = 'Phase')
DimPlot(myeloid, group.by = 'age')
FeaturePlot(myeloid, 'percent.mt')
FeaturePlot(myeloid, 'complexity')
```

```{r}
myeloid <- subset(myeloid, RNA_snn_res.0.1 != 3)
```


```{r include=FALSE}
myeloid <- myeloid %>% 
    NormalizeData() %>%
    FindVariableFeatures()%>%
    ScaleData() %>%
    RunPCA()

myeloid <- myeloid %>% 
    RunUMAP(dims = 1:5) %>%
    FindNeighbors(dims = 1:5) %>%
    FindClusters(resolution = c(seq(0.1, 1, 0.1), 2.0))

clustree(myeloid)
```

```{r eval=FALSE, include=FALSE}
DimPlot(myeloid, group.by = 'RNA_snn_res.0.5')
```

# SingleR

## clusters level

```{r echo=FALSE, fig.height=5, fig.width=8}
sce <- as.SingleCellExperiment(myeloid)
colLabels(sce) <- myeloid$RNA_snn_res.2

# #cluster-wise annotation
pred_bped_main <- SingleR(test = sce, ref = mouse_ref_imm, clusters=colLabels(sce),
                          labels = mouse_ref_imm$label.fine, BPPARAM=MulticoreParam(5))
# # Create a named vector for RenameIdents
pred_bped_main$pruned.labels <- gsub(" \\(.*\\)", "", pred_bped_main$pruned.labels)
singleR_labels <- pred_bped_main$pruned.labels
names(singleR_labels) <- rownames(pred_bped_main)

Idents(myeloid) <-  'RNA_snn_res.2'
# Update cluster identities in Seurat object based on SingleR results
myeloid <- RenameIdents(myeloid, singleR_labels)
myeloid[['SingleR_imm_ref_v1']] <- Idents(myeloid)

DimPlot(myeloid, group.by = 'SingleR_imm_ref_v1', cols = my_colors)
```

## single cell level

```{r echo=FALSE, fig.height=5, fig.width=8}

# #cluster-wise annotation
pred_bped_main <- SingleR(test = sce, ref = mouse_ref_imm,
                          labels = mouse_ref_imm$label.fine, BPPARAM=MulticoreParam(5))

pred_bped_main$pruned.labels <- gsub(" \\(.*\\)", "", pred_bped_main$pruned.labels)

myeloid[['SingleR_imm_ref_v2']] <- pred_bped_main$pruned.labels 

DimPlot(myeloid, group.by = 'SingleR_imm_ref_v2', cols = my_colors)
```

## Markers

### Macrophages

```{r fig.height=8, fig.width=10}
FeaturePlot(myeloid, c('C1qb', 'C1qa', 'Apoe', 'Itgam', 'Cd14', 'Cd68', 'Fcgr3', 'Fcgr1', 'Adgre1'))
```

### Monocytes

```{r fig.height=6, fig.width=10}
FeaturePlot(myeloid, c('Vcan', 'Thbs1', 'Ly6c2'))
```

### Dendritic cells

```{r fig.height=8, fig.width=10}
FeaturePlot(myeloid, c('Itgax', 'Ccl17', 'Cd86', 'Cd80', 'Itgae'))
```

### Neutrofils

```{r fig.height=4, fig.width=10}
FeaturePlot(myeloid, c('Cxcr2', 'Il1r2'))
```

```{r eval=FALSE, fig.height=6, fig.width=12, include=FALSE}
Idents(myeloid) <- 'RNA_snn_res.0.5'

clusters.markers <- FindAllMarkers(myeloid, only.pos = TRUE, min.pct = 0.25)
top5 <- clusters.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>% 
  # filter(!str_starts(gene, regex("^(Rb|Gm)")) & !str_ends(gene, "Rik")) %>%
  group_by(cluster) %>%
  top_n(n = 5, wt = avg_log2FC)

DotPlot(myeloid, top5$gene %>% unique(), cols = c('lightgrey', "#BA55D3FF"), scale = F) +
  RotatedAxis()

```

```{r eval=FALSE, include=FALSE}
DimPlot(myeloid, group.by = 'RNA_snn_res.0.9')
```


# Composition

```{r include=FALSE}
Idents(myeloid) <- 'RNA_snn_res.0.9'

new.cluster.ids <- c('0'='Macrophages',
                     '1'='DC',
                     '2'='Monocytes',
                     '3'='Macrophages',
                     '4'='Monocytes',
                     '5' = 'Macrophages',
                     '6' = 'Macrophages',
                     '7' = 'Macrophages',
                     '8' = 'DC',
                     '9' = 'Macrophages',
                     '10' = 'Neutrophils',
                     '11' = 'DC'
)

myeloid <- RenameIdents(myeloid, new.cluster.ids)
myeloid$annotation <- Idents(myeloid)
```

```{r echo=FALSE, fig.height=4, fig.width=12}
plot_grid(DimPlot(myeloid, group.by = 'annotation', cols = my_colors) +
  ggtitle("Annotation"), 
  DimPlot(myeloid, group.by = 'age', cols = c("#6495EDFF", "#FF69B4FF")) +
  ggtitle("Age"),
  nrow = 1)
```

```{r echo=FALSE, fig.height=4, fig.width=13}
plot_grid(dittoBarPlot(myeloid, 'age', group.by = 'annotation', color.panel = my_colors) +
            ggtitle('Age'),
          dittoBarPlot(myeloid, 'Phase', group.by = 'annotation', color.panel = my_colors),
          dittoBarPlot(myeloid, 'annotation', group.by = 'age', color.panel = my_colors) +
            ggtitle('Annotation'),
          rel_widths = c(1, 1, 0.7),
          nrow = 1)
```

```{r echo=FALSE, fig.height=6, fig.width=12, message=FALSE, warning=FALSE}
Idents(myeloid) <- 'annotation'

clusters.markers <- FindAllMarkers(myeloid, only.pos = TRUE, min.pct = 0.5, verbose = T)
top5 <- clusters.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>% 
  # filter(!str_starts(gene, regex("^(Rb|Gm)")) & !str_ends(gene, "Rik")) %>%
  group_by(cluster) %>%
  top_n(n = 7, wt = avg_log2FC)

DotPlot(myeloid, top5$gene %>% unique(), cols = c('lightgrey', "#BA55D3FF"), scale = F) +
  RotatedAxis()

```

```{r eval=FALSE, include=FALSE}
FeaturePlot(myeloid, c('Adgre1', 'Ptprc'))
FeaturePlot(myeloid, c('Cd163', 'Msr1', 'Il10', 'Cd68'))
FeaturePlot(myeloid, c('Ccl2', 'Vegfa', 'Il10', 'Cd68'))
DimPlot(myeloid, group.by = 'annotation', cols = my_colors) +
  ggtitle("Annotation")
```

# DEG by age

## Scatter plots

```{r echo=FALSE, fig.height=7, fig.width=7, message=FALSE, warning=FALSE}
myeloid$celltype.age <- paste(myeloid$annotation, myeloid$age, sep = "_")
Idents(myeloid) <- "celltype.age"
myeloid_bulk <- AggregateExpression(myeloid, group.by = c("annotation", "age"), return.seurat = TRUE)

p1 <- CellScatter(myeloid_bulk, "Macrophages_young", "Macrophages_adult")
p1 <- CellScatter(myeloid_bulk, "Macrophages_young", "Macrophages_adult", highlight = p1$data %>% filter(abs(`Macrophages_young` - `Macrophages_adult`) > 1) %>% rownames())
p1 <- LabelPoints(plot = p1, points = p1$data %>% filter(abs(`Macrophages_young` - `Macrophages_adult`) > 1) %>% rownames(), repel = TRUE)

p2 <- CellScatter(myeloid_bulk, "Monocytes_young", "Monocytes_adult")
p2 <- CellScatter(myeloid_bulk, "Monocytes_young", "Monocytes_adult", highlight = p2$data %>% filter(abs(`Monocytes_young` - `Monocytes_adult`) > 1) %>% rownames())
p2 <- LabelPoints(plot = p2, points = p2$data %>% filter(abs(`Monocytes_young` - `Monocytes_adult`) > 1) %>% rownames(), repel = TRUE)

p3 <- CellScatter(myeloid_bulk, "Neutrophils_young", "Neutrophils_adult")
p3 <- CellScatter(myeloid_bulk, "Neutrophils_young", "Neutrophils_adult", highlight = p3$data %>% filter(abs(Neutrophils_young - Neutrophils_adult) > 1) %>% rownames())
p3 <- LabelPoints(plot = p3, points = p3$data %>% filter(abs(Neutrophils_young - Neutrophils_adult) > 1) %>% rownames(), repel = TRUE)

p4 <- CellScatter(myeloid_bulk, "DC_young", "DC_adult")
p4 <- CellScatter(myeloid_bulk, "DC_young", "DC_adult", highlight = p4$data %>% filter(abs(`DC_young` - `DC_adult`) > 1) %>% rownames())
p4 <- LabelPoints(plot = p4, points =  p4$data %>% filter(abs(`DC_young` - `DC_adult`) > 1) %>% rownames(), repel = TRUE)

p1 + p2 + p3 + p4
```


```{r eval=FALSE, fig.height=6, fig.width=12, include=FALSE}
genes <- c(p1$data %>% filter(abs(`Macrophages_young` - `Macrophages_adult`) > 1) %>% rownames(),
           p4$data %>% filter(abs(`DC_young` - `DC_adult`) > 1) %>% rownames(),
           p2$data %>% filter(abs(`Monocytes_young` - `Monocytes_adult`) > 1) %>% rownames()
           ) %>% unique
Idents(myeloid) <- 'annotation'
DotPlot(myeloid, genes, cols = c("#6495EDFF", "#FF69B4FF"), dot.scale = 8, 
    split.by = "age") + RotatedAxis()
```

## Volcano plots

```{r echo=FALSE, fig.height=8, fig.width=10}
macr.de.markers <- FindMarkers(myeloid, 
                                  ident.1 = "Macrophages_young", ident.2 = "Macrophages_adult", 
                                  min.pct = 0.25)
mono.de.markers <- FindMarkers(myeloid,
                                  ident.1 = "Monocytes_young", ident.2 = "Monocytes_young",
                                  min.pct = 0.01)
dc.de.markers <- FindMarkers(myeloid, 
                                  ident.1 = "DC_young", ident.2 = "DC_adult", 
                                  min.pct = 0.25)

volcano_plot <- function(de.markers) {
  EnhancedVolcano(de.markers,
    lab = rownames(de.markers),
    x = 'avg_log2FC',
    y = 'p_val_adj',
    col = c("grey30","#8FBC8BFF", "#6495EDFF", "#BA55D3FF"),
    drawConnectors = TRUE,
    pCutoff = 0.05, 
    FCcutoff = 0.5,
    max.overlaps = 30
    ) +
  labs(subtitle = "Young vs Adult")
}

plot_grid(volcano_plot(macr.de.markers) +
            ggtitle('Macrophages'),
          # volcano_plot(mono.de.markers) +
          #   ggtitle('Monocytes'),
          volcano_plot(dc.de.markers) +
            ggtitle('DC'),
          nrow = 1)
```

```{r}
Idents(myeloid) <- 'age'
VlnPlot(subset(myeloid, annotation == 'Macrophages'), 'Apoe')
FeaturePlot(myeloid, 'Apoe')
```

