---
title: "Myeloid_cells_analysis"
output: html_document
date: "2024-08-24"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = FALSE,
	message = FALSE,
	warning = FALSE
)
library(data.table)
library(Seurat)
library(scDblFinder)
library(DoubletFinder)
library(dplyr)
library(remotes)
# remotes::install_github("carmonalab/STACAS")
library(STACAS)
library(SingleR)
# install.packages("scGate")
library(scGate)
library(ggplot2)
library(BiocParallel)
library(harmony)
library(RColorBrewer)
library(stringr)
library(cowplot)
library(scuttle)
# install.packages("devtools")
# BiocManager::install("Homo.sapiens")
# BiocManager::install("jpmam1/scalop") #forked copy of a really good guy 
# devtools::install_github("jlaffy/infercna")
library(infercna)
# library(infercnv)
library(parallel)
library(future)
library(clustree)
library(dittoSeq)
# install.packages("ggh4x")
library(ggh4x)
#BiocManager::install('EnhancedVolcano')
library(EnhancedVolcano)
library(clusterProfiler)
# BiocManager::install('org.Mm.eg.db')
library("org.Mm.eg.db")
library(ReactomePA)
library(msigdbr)
# install.packages('qs')
library(qs)
# devtools::install_github("ncborcherding/escape")
# devtools::install_github("rcastelo/GSVA")
library(escape)
# install.packages('gprofiler2')
library(gprofiler2)
# remotes::install_github("carmonalab/ProjecTILs")
library(ProjecTILs)
library(gridExtra)
```

```{r global variables, include=FALSE}
set.seed(100)
DATADIR = '~/Desktop/lab/neuroblastoma/data/'
WORKDIR = '~/Desktop/lab/neuroblastoma/neuroblastoma/'
PREPRDATADIR = '~/Desktop/lab/neuroblastoma/data/'

my_colors <- c("#9ACD32FF", "#4682B4FF",  "#DDA0DDFF", "#FFA07AFF", "#8FBC8BFF",
               "#40E0D0FF",  "#F0E68CFF", "#5F9EA0FF", "#D2B48CFF",  
               "#6495EDFF", "#FF69B4FF", "#BA55D3FF",  "#F08080FF", "#32CD32FF",  
               "#FFDAB9FF",  "#87CEEBFF")
```

```{r global variables server, eval=FALSE, include=FALSE}
set.seed(100)
DATADIR = '~/neuroblastoma/data/'
WORKDIR = '~/neuroblastoma/'
PREPRDATADIR = '~/neuroblastoma/preprocessing_data/'

my_colors <- c("#9ACD32FF", "#4682B4FF",  "#DDA0DDFF", "#FFA07AFF", "#8FBC8BFF",
               "#40E0D0FF",  "#F0E68CFF", "#5F9EA0FF", "#D2B48CFF",  
               "#6495EDFF", "#FF69B4FF", "#BA55D3FF",  "#F08080FF", "#32CD32FF",  
               "#FFDAB9FF",  "#87CEEBFF")
```

```{r download references, message=FALSE, warning=FALSE, include=FALSE}
mouse_ref <- celldex::MouseRNAseqData()
mouse_ref_imm <- celldex::ImmGenData()
scGate_models_DB <- get_scGateDB()
```

```{r}
seu <- qread(paste0(PREPRDATADIR, 'seu_mouse.qs'))
```

```{r}
myeloid <- subset(seu, annotation == 'myeloid_cells')
```

```{r include=FALSE}
mmus_s = gorth(cc.genes$s.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name
mmus_g2m = gorth(cc.genes$g2m.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name

myeloid <- CellCycleScoring(myeloid, s.features = mmus_s, g2m.features = mmus_g2m)
```

```{r}
myeloid <- myeloid %>% 
    NormalizeData() %>%
    FindVariableFeatures()%>%
    ScaleData() %>%
    RunPCA()

myeloid <- myeloid %>% 
    RunUMAP(dims = 1:4) %>%
    FindNeighbors(dims = 1:4) %>%
    FindClusters(resolution = seq(0.1, 1, 0.1))
```

```{r fig.height=20, fig.width=10}
DimHeatmap(myeloid, dims = 1:20)
```

```{r}
ElbowPlot(myeloid)
```

```{r}
clustree(myeloid)
```

```{r}
DimPlot(myeloid, group.by = 'RNA_snn_res.0.1')
DimPlot(myeloid, group.by = 'Phase')
DimPlot(myeloid, group.by = 'age')
FeaturePlot(myeloid, 'percent.mt')
FeaturePlot(myeloid, 'complexity')
```

```{r}
myeloid <- subset(myeloid, RNA_snn_res.0.1 != 3)
```


```{r}
myeloid <- myeloid %>% 
    NormalizeData() %>%
    FindVariableFeatures()%>%
    ScaleData() %>%
    RunPCA()

myeloid <- myeloid %>% 
    RunUMAP(dims = 1:5) %>%
    FindNeighbors(dims = 1:5) %>%
    FindClusters(resolution = seq(0.1, 1, 0.1))

clustree(myeloid)
```

```{r}
DimPlot(myeloid, group.by = 'RNA_snn_res.0.5')
```

# SingleR

## Immune reference from celldex

```{r fig.width=10, include=FALSE}
sce <- as.SingleCellExperiment(myeloid)
colLabels(sce) <- myeloid$RNA_snn_res.1

# #cluster-wise annotation
pred_bped_main <- SingleR(test = sce, ref = mouse_ref_imm, clusters=colLabels(sce),
                          labels = mouse_ref_imm$label.fine, BPPARAM=MulticoreParam(5))
# # Create a named vector for RenameIdents
singleR_labels <- pred_bped_main$pruned.labels
names(singleR_labels) <- rownames(pred_bped_main)

Idents(myeloid) <-  'RNA_snn_res.1'
# Update cluster identities in Seurat object based on SingleR results
myeloid <- RenameIdents(myeloid, singleR_labels)
myeloid[['SingleR_imm_ref_v1']] <- Idents(myeloid)

DimPlot(myeloid, group.by = 'SingleR_imm_ref_v1', cols = my_colors)
```

## Immune reference from celldex

```{r fig.height=6, fig.width=18, include=FALSE}

# #cluster-wise annotation
pred_bped_main <- SingleR(test = sce, ref = mouse_ref_imm,
                          labels = mouse_ref_imm$label.fine, BPPARAM=MulticoreParam(5))

myeloid[['SingleR_imm_ref_v2']] <- pred_bped_main$pruned.labels

DimPlot(myeloid, group.by = 'SingleR_imm_ref_v2')
```

# scGate annotation

```{r}
myeloid <- scGate(myeloid, model = scGate_models_DB$mouse$generic$MoMacDC)
DimPlot(myeloid, group.by = 'scGate_multi')
FeaturePlot(myeloid, 'Epithelial_UCell')
```

## Markers

```{r fig.height=6, fig.width=12}
Idents(myeloid) <- 'RNA_snn_res.0.5'

clusters.markers <- FindAllMarkers(myeloid, only.pos = TRUE, min.pct = 0.25)
top5 <- clusters.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>% 
  # filter(!str_starts(gene, regex("^(Rb|Gm)")) & !str_ends(gene, "Rik")) %>%
  group_by(cluster) %>%
  top_n(n = 5, wt = avg_log2FC)

DotPlot(myeloid, top5$gene %>% unique(), cols = c('lightgrey', "#BA55D3FF"), scale = F) +
  RotatedAxis()

```

0 -
1 - macrofages
2 - 
3 - 
4 - 

```{r}
clusters.markers %>% 
  dplyr::filter(p_val_adj < 0.05) %>% 
  dplyr::filter(cluster == 3) %>% 
  filter(!str_starts(gene, regex("^(Rb|Gm)")) & !str_ends(gene, "Rik"))  %>% 
  top_n(n = 20, wt = avg_log2FC) %>% 
  rownames()
```


```{r}
FeaturePlot(myeloid, 'Mycn')

mouse_ref_imm$label.fine %>% unique()
```


```{r}
DimPlot(seu, group.by = 'annotation')
```

