---
title: "Myeloid_cells_analysis"
output: html_document
date: "2024-08-24"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = FALSE,
	message = FALSE,
	warning = FALSE
)
library(data.table)
library(Seurat)
library(scDblFinder)
library(DoubletFinder)
library(dplyr)
library(remotes)
# remotes::install_github("carmonalab/STACAS")
library(STACAS)
library(SingleR)
# install.packages("scGate")
library(scGate)
library(ggplot2)
library(BiocParallel)
library(harmony)
library(RColorBrewer)
library(stringr)
library(cowplot)
library(scuttle)
# install.packages("devtools")
# BiocManager::install("Homo.sapiens")
# BiocManager::install("jpmam1/scalop") #forked copy of a really good guy 
# devtools::install_github("jlaffy/infercna")
library(infercna)
# library(infercnv)
library(parallel)
library(future)
library(clustree)
library(dittoSeq)
# install.packages("ggh4x")
library(ggh4x)
#BiocManager::install('EnhancedVolcano')
library(EnhancedVolcano)
library(clusterProfiler)
# BiocManager::install('org.Mm.eg.db')
library("org.Mm.eg.db")
library(ReactomePA)
library(msigdbr)
# install.packages('qs')
library(qs)
# devtools::install_github("ncborcherding/escape")
# devtools::install_github("rcastelo/GSVA")
library(escape)
# install.packages('gprofiler2')
library(gprofiler2)
# remotes::install_github("carmonalab/ProjecTILs")
library(ProjecTILs)
library(gridExtra)
library(scCustomize)
library(enrichR)
# devtools::install_github("jinworks/CellChat")
```

```{r global variables server, include=FALSE}
set.seed(100)
DATADIR = '~/neuroblastoma/data/'
WORKDIR = '~/neuroblastoma/'
PREPRDATADIR = '~/neuroblastoma/preprocessing_data/'

my_colors <- c("#9ACD32FF", "#4682B4FF",  "#DDA0DDFF", "#FFA07AFF", "#8FBC8BFF",
               "#40E0D0FF",  "#F0E68CFF", "#5F9EA0FF", "#D2B48CFF",  
               "#6495EDFF", "#FF69B4FF", "#BA55D3FF",  "#F08080FF", "#32CD32FF",  
               "#FFDAB9FF",  "#87CEEBFF")
```

```{r}
volcano_plot <- function(de.markers) {
  EnhancedVolcano(de.markers,
    lab = rownames(de.markers),
    x = 'avg_log2FC',
    y = 'p_val_adj',
    col = c("grey30","#8FBC8BFF", "#6495EDFF", "#BA55D3FF"),
    drawConnectors = TRUE,
    pCutoff = 0.05, 
    FCcutoff = 0.5,
    max.overlaps = 30
    ) +
  labs(subtitle = "Young vs Adult")
}
```

```{r download references, message=FALSE, warning=FALSE, include=FALSE}
mouse_ref <- celldex::MouseRNAseqData()
mouse_ref_imm <- celldex::ImmGenData()
scGate_models_DB <- get_scGateDB()
```

```{r include=FALSE}
seu <- qread(paste0(PREPRDATADIR, 'seu_mouse.qs'))
```

```{r include=FALSE}
myeloid <- subset(seu, annotation == 'myeloid_cells')
```

```{r include=FALSE}
mmus_s = gorth(cc.genes$s.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name
mmus_g2m = gorth(cc.genes$g2m.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name

myeloid <- CellCycleScoring(myeloid, s.features = mmus_s, g2m.features = mmus_g2m)
```

```{r include=FALSE}
myeloid <- myeloid %>% 
    NormalizeData() %>%
    FindVariableFeatures()%>%
    ScaleData() %>%
    RunPCA()

myeloid <- myeloid %>% 
    RunUMAP(dims = 1:4) %>%
    FindNeighbors(dims = 1:4) %>%
    FindClusters(resolution = seq(0.1, 1, 0.1))
```

```{r eval=FALSE, fig.height=20, fig.width=10, include=FALSE}
DimHeatmap(myeloid, dims = 1:20)
```

```{r eval=FALSE, include=FALSE}
ElbowPlot(myeloid)
```

```{r include=FALSE}
DimPlot(myeloid, group.by = 'RNA_snn_res.0.1')
DimPlot(myeloid, group.by = 'Phase')
DimPlot(myeloid, group.by = 'age')
FeaturePlot(myeloid, 'percent.mt')
FeaturePlot(myeloid, 'complexity')
```

```{r}
myeloid <- subset(myeloid, RNA_snn_res.0.1 != 3)
```


```{r include=FALSE}
myeloid <- myeloid %>% 
    NormalizeData() %>%
    FindVariableFeatures()%>%
    ScaleData() %>%
    RunPCA()

myeloid <- myeloid %>% 
    RunUMAP(dims = 1:5) %>%
    FindNeighbors(dims = 1:5) %>%
    FindClusters(resolution = c(seq(0.1, 1, 0.1), 2.0))

clustree(myeloid)
```

```{r eval=FALSE, include=FALSE}
DimPlot(myeloid, group.by = 'RNA_snn_res.0.5')
```

# SingleR

## clusters level

```{r echo=FALSE, fig.height=5, fig.width=8}
sce <- as.SingleCellExperiment(myeloid)
colLabels(sce) <- myeloid$RNA_snn_res.2

# #cluster-wise annotation
pred_bped_main <- SingleR(test = sce, ref = mouse_ref_imm, clusters=colLabels(sce),
                          labels = mouse_ref_imm$label.fine, BPPARAM=MulticoreParam(5))
# # Create a named vector for RenameIdents
pred_bped_main$pruned.labels <- gsub(" \\(.*\\)", "", pred_bped_main$pruned.labels)
singleR_labels <- pred_bped_main$pruned.labels
names(singleR_labels) <- rownames(pred_bped_main)

Idents(myeloid) <-  'RNA_snn_res.2'
# Update cluster identities in Seurat object based on SingleR results
myeloid <- RenameIdents(myeloid, singleR_labels)
myeloid[['SingleR_imm_ref_v1']] <- Idents(myeloid)

DimPlot(myeloid, group.by = 'SingleR_imm_ref_v1', cols = my_colors)
```

## single cell level

```{r echo=FALSE, fig.height=5, fig.width=8}

# #cluster-wise annotation
pred_bped_main <- SingleR(test = sce, ref = mouse_ref_imm,
                          labels = mouse_ref_imm$label.fine, BPPARAM=MulticoreParam(5))

pred_bped_main$pruned.labels <- gsub(" \\(.*\\)", "", pred_bped_main$pruned.labels)

myeloid[['SingleR_imm_ref_v2']] <- pred_bped_main$pruned.labels 

DimPlot(myeloid, group.by = 'SingleR_imm_ref_v2', cols = my_colors)
```

## Markers

### Macrophages

```{r fig.height=8, fig.width=10}
FeaturePlot(myeloid, c('C1qb', 'C1qa', 'Apoe', 'Itgam', 'Cd14', 'Cd68', 'Fcgr3', 'Fcgr1', 'Adgre1'))
```

### Monocytes

```{r fig.height=6, fig.width=10}
FeaturePlot(myeloid, c('Vcan', 'Thbs1', 'Ly6c2'))
```

### Dendritic cells

```{r fig.height=8, fig.width=10}
FeaturePlot(myeloid, c('Itgax', 'Ccl17', 'Cd86', 'Cd80', 'Itgae'))
```

### Neutrofils

```{r fig.height=4, fig.width=10}
FeaturePlot(myeloid, c('Cxcr2', 'Il1r2'))
```

```{r eval=FALSE, fig.height=6, fig.width=15, include=FALSE}
Idents(myeloid) <- 'RNA_snn_res.0.9'

clusters.markers <- FindAllMarkers(myeloid, only.pos = TRUE, min.pct = 0.25)
top5 <- clusters.markers %>%
  dplyr::filter(p_val_adj < 0.05) %>% 
  # filter(!str_starts(gene, regex("^(Rb|Gm)")) & !str_ends(gene, "Rik")) %>%
  group_by(cluster) %>%
  top_n(n = 5, wt = avg_log2FC)

DotPlot(myeloid, c('Adgre1','Itgam', 'Cd68', 'Mrc1', 'Ccr7', 'Flt3', 'Maf', top5$gene %>% unique()), cols = c('lightgrey', "#BA55D3FF"), scale = F) +
  RotatedAxis()

```


```{r eval=FALSE, include=FALSE}
DimPlot(myeloid, group.by = 'RNA_snn_res.0.9')
```


# Composition

```{r eval=FALSE, include=FALSE}
Idents(myeloid) <- 'RNA_snn_res.0.9'

new.cluster.ids <- c('0'='Macrophages', 
                     '1'='DC',
                     '2'='Monocytes',
                     '3'='Macrophages',
                     '4'='Monocytes',
                     '5' = 'Macrophages',
                     '6' = 'Macrophages',
                     '7' = 'Macrophages',
                     '8' = 'DC',
                     '9' = 'Macrophages',
                     '10' = 'Neutrophils',
                     '11' = 'DC'
)

myeloid <- RenameIdents(myeloid, new.cluster.ids)
myeloid$annotation <- Idents(myeloid)
```

```{r include=FALSE}
Idents(myeloid) <- 'RNA_snn_res.0.9'

new.cluster.ids <- c('0'='TAMs Maf+', 
                     '1'='cDC',
                     '2'='Monocytes Ly6c+',
                     '3'='TAMs Maf+',
                     '4'='Monocytes Ly6c+Nr4a1+',
                     '5' = 'TAMs Maf+',
                     '6' = 'TAMs Ero1a+',
                     '7' = 'TAMs Cmss1+Ftl+Lgals3+',
                     '8' = 'cDC Ccr7+',
                     '9' = 'TAMs Prg4+Alox15+',
                     '10' = 'Neutrophils',
                     '11' = 'pDC'
)

myeloid <- RenameIdents(myeloid, new.cluster.ids)
myeloid$annotation <- Idents(myeloid)
```


```{r echo=FALSE, fig.height=4, fig.width=12}
custom_order <- c("Monocytes Ly6c+", "Monocytes Ly6c+Nr4a1+",  "Neutrophils",
                  'TAMs Cmss1+Ftl+Lgals3+',
                  'TAMs Maf+', "TAMs Ero1a+", 'TAMs Prg4+Alox15+',
                   'cDC', 'cDC Ccr7+', 'pDC')
myeloid$annotation <- factor(myeloid$annotation, levels = custom_order)

plot_grid(DimPlot(myeloid, group.by = 'annotation', cols = my_colors) +
  ggtitle("Annotation"), 
  DimPlot(myeloid, group.by = 'age', cols = c("#6495EDFF", "#FF69B4FF")) +
  ggtitle("Age"),
  nrow = 1, rel_widths = c(1, 0.8))
```

```{r echo=FALSE, fig.height=5, fig.width=14}
plot_grid(dittoBarPlot(myeloid, 'age', group.by = 'annotation', color.panel = my_colors, scale = 'count') +
            ggtitle('Age'),
          dittoBarPlot(myeloid, 'Phase', group.by = 'annotation', color.panel = my_colors),
          dittoBarPlot(myeloid, 'annotation', group.by = 'age', color.panel = my_colors) +
            ggtitle('Annotation'),
          rel_widths = c(1, 1, 0.7),
          nrow = 1)
```

We see that in TME of neuroblastoma of adult mouse there are more Monocytes Ly6c(high), more TAMs Ero1a+ (found only in adult), TAMs Maf+, while in TME of neuroblastoma of young mouse there are more Monocytes Ly6c(high)Nr4a1+ (relatively), pDC and cDC (relatively), TAMs Cmss1+Ftl+Lgals3+ (relatively) and neutrophils (relatively)



```{r echo=FALSE, fig.height=8, fig.width=7}
Idents(myeloid) <- 'annotation'

clusters.markers <- FindAllMarkers(myeloid, only.pos = TRUE, min.pct = 0.25)
top5 <- clusters.markers %>%
  dplyr::filter(p_val_adj < 0.05 & pct.1 > 0.45) %>% 
  # filter(!str_starts(gene, regex("^(Rb|Gm)")) & !str_ends(gene, "Rik")) %>%
  group_by(cluster) %>%
  top_n(n = 5, wt = avg_log2FC) %>% 
  rbind(clusters.markers %>%
  dplyr::filter(gene == 'Maf' | gene == 'H2-Ob'))

custom_order <- c("Neutrophils", "TAMs Ero1a+", "Monocytes Ly6c+Nr4a1+", "Monocytes Ly6c+",
                  'TAMs Maf+', 'TAMs Cmss1+Ftl+Lgals3+', 'TAMs Prg4+Alox15+',  'cDC', 
                  'cDC Ccr7+', 'pDC')

# Convert the 'name' column to a factor with a specified order
top5$cluster <- factor(top5$cluster, levels = custom_order)
top5 <- top5[order(top5$cluster), ]

scCustomize::Clustered_DotPlot(myeloid, features = c(top5$gene[1:46],'Siglech', 'Ly6d', 'Ccr9', 'Bst2', 'Itgax') %>% unique(),
                               plot_km_elbow = FALSE, cluster_feature = F,
                               colors_use_idents = my_colors,
                               raster = T)
```

# DEG

```{r fig.height=8, fig.width=8}
Idents(myeloid) <- 'age'
de.markers <- FindMarkers(myeloid, 
                                  ident.1 = "young", ident.2 = "adult", 
                                  min.pct = 0.15)
de.markers <- de.markers %>% 
  mutate(gene = rownames(.)) %>% 
  dplyr::filter(!str_starts(gene, regex("^(Rb|Gm|Hb)")) & !str_ends(gene, "Rik")) %>% 
  dplyr::filter(gene != 'Igf2bp3')

plot_grid(volcano_plot(de.markers) +
            ggtitle('Myeloid cells'),
          nrow = 1)
```

# Human validation

```{r include=FALSE}
myeloid_h <- qread(paste0(PREPRDATADIR, 'myeloid_cells_v3.qs'))
```

```{r eval=FALSE, include=FALSE}
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes 
myeloid_h <- myeloid_h %>% 
  CellCycleScoring(s.features = s.genes, g2m.features = g2m.genes) 
myeloid_h[["RNA"]] <- split(myeloid_h[["RNA"]], f = myeloid_h$SampleID)

myeloid_h <- myeloid_h %>% 
  NormalizeData() %>% 
  FindVariableFeatures() %>% 
  ScaleData(vars.to.regress = c("S.Score", "G2M.Score"), verbose = F) %>% 
  RunPCA(verbose=FALSE)
  
myeloid_h <- myeloid_h %>% 
  IntegrateLayers(
    object = ., method = CCAIntegration,
    orig.reduction = "pca", new.reduction = "integrated.cca.myeloid",
    verbose = FALSE, k.weight = min(table(myeloid_h$SampleID))) 

myeloid <- JoinLayers(myeloid)

qsave(myeloid_h, paste0(PREPRDATADIR, 'myeloid_cells_v3.qs'))
```

```{r include=FALSE}
myeloid_h <- myeloid_h %>% 
  RunUMAP(reduction = "integrated.cca.myeloid", dims = 1:8) %>% 
  FindNeighbors(dims = 1:8) %>%
  FindClusters(resolution = seq(0.1, 1, 0.1))


clustree(myeloid_h)
```

```{r eval=FALSE, fig.height=20, fig.width=12, include=FALSE}
DimHeatmap(myeloid_h, dims = 1:20)
```

```{r eval=FALSE, include=FALSE}
ElbowPlot(myeloid_h)
```

```{r eval=FALSE, include=FALSE}
DimPlot(myeloid_h, group.by = 'RNA_snn_res.0.1')
# replace_list <- c('Fibroblasts', "Plasma cells", "MEP", "GMP", "Neurons", "Neutrophils", "Endothelial cells", "naive B-cells")
# t_cells_h$SingleR_fine_tune <- ifelse(t_cells_h$SingleR_fine %in% replace_list, NA, t_cells_h$SingleR_fine)
# DimPlot(myeloid_h, group.by = 'SingleR_fine')
# DimPlot(myeloid_h, group.by = 'SingleR_fine_tune', cols = my_colors)
DimPlot(myeloid_h, group.by = 'celltype_humatlas_main')
DimPlot(myeloid_h, group.by = 'scGate_multi', cols = my_colors)
DimPlot(myeloid_h, group.by = 'Phase')
DimPlot(myeloid_h, group.by = 'Study')

```

## ProjecTILs

### Mouse reference

```{r include=FALSE}
ref.mouse <- make.reference(ref = myeloid, ndim = 5, seed = 100, recalculate.umap = FALSE,
    annotation.column = "annotation")
query.projected <- Run.ProjecTILs(myeloid_h, ref = ref.mouse, filter.cell = F, skip.normalize = T, split.by = 'SampleID', ncores = 14)
```

```{r echo=FALSE}
plot.projection(ref.mouse, query.projected, linesize = 0.3, pointsize = 0.3)
```

## Markers

```{r fig.height=14, fig.width=19}
DefaultAssay(object = query.projected) <- "integrated"
query.list <- SplitObject(query.projected, split.by = "less18M")
names(query.list) <- c('adult', 'young')


genes <- c(top5$gene[1:46],'Siglech', 'Ly6d', 'Ccr9', 'Bst2', 'Itgax') %>% unique()
genes <- genes[c(top5$gene[1:46],'Siglech', 'Ly6d', 'Ccr9', 'Bst2', 'Itgax') %>% unique() %in% Features(query.list$adult) & c(top5$gene[1:46],'Siglech', 'Ly6d', 'Ccr9', 'Bst2', 'Itgax') %>% unique() %in% Features(query.list$young) & !(c(top5$gene[1:46],'Siglech', 'Ly6d', 'Ccr9', 'Bst2', 'Itgax') %>% unique() %in% c('Ptgis', 'Rp1', 'Vmn2r26', "Tmem150c", "Lad1", "Slco5a1","Adcy6", "Slc2a1", "Dusp2", "Dusp16", "Ms4a7", "H2-Ob", "Dpp4", "Alox15", 'Prg4', "Ly6d", "Bst2"))] 

plot.states.radar(ref.mouse, query = query.list, ref.assay = 'integrated', query.assay = 'integrated', min.cells = 5, genes4radar = genes)
```

## Comparison

```{r include=FALSE}
p1 <- plot.projection(ref.mouse, query.list$young, linesize = 0.3, pointsize = 0.3) + ggtitle('Young (< 18 months)')
p2 <- plot.projection(ref.mouse, query.list$adult, linesize = 0.3, pointsize = 0.3)  + ggtitle('Adult (> 18 months)')
```

```{r echo=FALSE, fig.height=9, fig.width=16}
plot_grid(p1, p2, nrow = 1)
```

```{r include=FALSE}
# Step 1: Combine 'young' cluster information
young_cluster_info <- query.list[["young"]]$functional.cluster
young_metadata_vector <- as.character(young_cluster_info)
names(young_metadata_vector) <- names(young_cluster_info)

# Initialize the vector with NA for all cells
projection_annotation <- rep(NA, length(Cells(myeloid_h)))
names(projection_annotation) <- Cells(myeloid_h)

# Assign 'young' annotations
projection_annotation[names(young_metadata_vector)] <- young_metadata_vector

# Step 2: Combine 'adult' cluster information
adult_cluster_info <- query.list[["adult"]]$functional.cluster
adult_metadata_vector <- as.character(adult_cluster_info)
names(adult_metadata_vector) <- names(adult_cluster_info)

# Assign 'adult' annotations, keeping existing 'young' annotations where they exist
projection_annotation[names(adult_metadata_vector)] <- ifelse(
  is.na(projection_annotation[names(adult_metadata_vector)]),
  adult_metadata_vector,
  projection_annotation[names(adult_metadata_vector)]
)


# Add the final combined annotation to the Seurat object
myeloid_h$projection_annotation <- factor(projection_annotation, 
                                        levels = c("Monocytes Ly6c+","Monocytes Ly6c+Nr4a1+", 
                                                   "Neutrophils",'TAMs Cmss1+Ftl+Lgals3+',
                                                   'TAMs Maf+', "TAMs Ero1a+", 'TAMs Prg4+Alox15+',
                                                   'cDC', 'cDC Ccr7+', 'pDC'))
```

## Composition

```{r echo=FALSE, fig.height=4, fig.width=12}
plot_grid(DimPlot(myeloid_h, group.by = 'projection_annotation', cols = my_colors) +
  ggtitle("Annotation"), 
  DimPlot(myeloid_h, group.by = 'less18M', cols = c("#6495EDFF", "#FF69B4FF")) +
  ggtitle("Age"),
  nrow = 1, rel_widths = c(1, 0.8))
```

```{r echo=FALSE, fig.height=5, fig.width=14}
plot_grid(dittoBarPlot(myeloid_h, 'less18M', group.by = 'projection_annotation', color.panel = my_colors, scale = 'count') +
            ggtitle('Age'),
          dittoBarPlot(myeloid_h, 'Phase', group.by = 'projection_annotation', color.panel = my_colors),
          dittoBarPlot(myeloid_h, 'projection_annotation', group.by = 'less18M', color.panel = my_colors) +
            ggtitle('Annotation'),
          rel_widths = c(1, 1, 0.7),
          nrow = 1)
```

### scGate 

```{r echo=FALSE, fig.height=5, fig.width=14}
plot_grid(dittoBarPlot(myeloid_h, 'less18M', group.by = 'scGate_multi', color.panel = my_colors, scale = 'count') +
            ggtitle('Age'),
          dittoBarPlot(myeloid_h, 'Phase', group.by = 'scGate_multi', color.panel = my_colors),
          dittoBarPlot(myeloid_h, 'scGate_multi', group.by = 'less18M', color.panel = my_colors) +
            ggtitle('Annotation'),
          rel_widths = c(1, 1, 0.7),
          nrow = 1)
```
# Mouse vs Human

```{r}
plot_grid(dittoBarPlot(myeloid_h, 'projection_annotation', group.by = 'less18M', color.panel = my_colors) +
            ggtitle('Human'),
          dittoBarPlot(myeloid, 'annotation', group.by = 'age', color.panel = my_colors) +
            ggtitle('Mouse'))
```

